<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Biostatistics - Linear model extensions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../content/chapters/multivariate_methods.html" rel="next">
<link href="../../content/chapters/Combining_numerical_and_categorical_predictors.html" rel="prev">
<link href="../../images/favicon.ico" rel="icon">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../content/chapters/Introduction.html">Chapters</a></li><li class="breadcrumb-item"><a href="../../content/chapters/Linear_model_extensions.html">Extending the linear model</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../../index.html" class="sidebar-logo-link">
      <img src="../../images/image-from-rawpixel-id-547642-jpeg.jpg" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">An introduction to statistics and modern tools</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://www.gosnelllab.com/" title="Gosnell lab website" class="quarto-navigation-tool px-1" aria-label="Gosnell lab website"><i class="bi bi-globe"></i></a>
    <a href="https://github.com/jsgosnell" title="J. Stephen Gosnell github" class="quarto-navigation-tool px-1" aria-label="J. Stephen Gosnell github"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Getting started</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/getting_started/getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Before the first class</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Chapters</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Acquiring_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acquiring data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/summarizing_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Summarizing data</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Estimation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Estimation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Probability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Probability</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Binomial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hypothesis testing starting with binomial tests</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Tests_for_continuous_data_from_one_sample.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tests for continuous data from one sample</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Compare_proportions_among_populations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Compare proportions among groups</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Compare_means_among_populations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Compare means among groups</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/More_ANOVAs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More ANOVAs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Relationships_among_numerical_variables.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Relationships among numerical variables</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Combining_numerical_and_categorical_predictors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Combining numerical and categorical predictors</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/Linear_model_extensions.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Extending the linear model</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/multivariate_methods.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multivariate methods</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/chapters/bayesian_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bayesian analysis</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Extensions</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/extensions/maps.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Maps</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">swirl lessons</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/swirl_lessons.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">swirl lessons</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Practice problems</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/how_to_practice.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">How to practice</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/1_Getting_used_to_R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Getting used to R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/2_Estimates_and_ggplot2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Estimates and ggplot2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/3_Introduction_to_hypothesis_testing_via_binomial_test.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hypothesis testing starting with binomial tests</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/4_Continuous_tests_for_1_population.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tests for continuous data from one sample</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/5_Contingency_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Compare proportions among groups</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/6_Compare_means.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Compare means among groups</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/7_More_ANOVAs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More ANOVAs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/8_Relationships_among_numerical_variables.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Relationships among numerical variables</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/9_Combining_numerical_and_categorical_predictors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Combining numerical and categorical predictors</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/10_Linear_model_extensions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Linear model extensions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/practice_problems/11_Multivariate_methods.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multivariate methods</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false">
 <span class="menu-text">Solutions</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/1_Getting_used_to_R_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Getting used to R</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/2_Estimates_and_ggplot2_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Estimates and ggplot2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/3_Introduction_to_hypothesis_testing_via_binomial_test_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hypothesis testing starting with binomial tests</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/4_Continuous_tests_for_1_population_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tests for continuous data from one sample</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/5_Contingency_analysis_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Compare proportions among groups</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/6_Compare_means_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Compare means among groups</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/7_More_ANOVAs_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More ANOVAs</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/8_Relationships_among_numerical_variables_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Relationships among numerical variables</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/9_Combining_numerical_and_categorical_predictors_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Combining numerical and categorical predictors</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/10_Linear_model_extensions_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Linear model extensions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/solutions/11_Multivariate_methods_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multivariate methods</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false">
 <span class="menu-text">Software primer</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/primer_materials/tools_overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tools overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/primer_materials/R_primer.R" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Download an R primer file!</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="false">
 <span class="menu-text">Appendix (Odds and ends)</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/Appendix/github_classroom_stuff.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Github classroom stuff</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" aria-expanded="false">
 <span class="menu-text">End matter</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/end_matter/acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/end_matter/additional_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additional resources</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/references.bib" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Download a full reference list!</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#sticking-with-the-linear-model" id="toc-sticking-with-the-linear-model" class="nav-link active" data-scroll-target="#sticking-with-the-linear-model">Sticking with the linear model</a></li>
  <li><a href="#broken-assumption-linear-relationship-is-inappropriate" id="toc-broken-assumption-linear-relationship-is-inappropriate" class="nav-link" data-scroll-target="#broken-assumption-linear-relationship-is-inappropriate">Broken assumption: Linear relationship is inappropriate</a>
  <ul>
  <li><a href="#transform-the-data-least-advisable-sometimes" id="toc-transform-the-data-least-advisable-sometimes" class="nav-link" data-scroll-target="#transform-the-data-least-advisable-sometimes">Transform the data (least advisable, sometimes)</a></li>
  <li><a href="#different-data-require-a-different-approach-introducing-generalized-linear-models" id="toc-different-data-require-a-different-approach-introducing-generalized-linear-models" class="nav-link" data-scroll-target="#different-data-require-a-different-approach-introducing-generalized-linear-models">Different data require a different approach: Introducing generalized linear models</a>
  <ul>
  <li><a href="#logistic-regression-bernoulli-and-binomial-outcomes" id="toc-logistic-regression-bernoulli-and-binomial-outcomes" class="nav-link" data-scroll-target="#logistic-regression-bernoulli-and-binomial-outcomes">Logistic regression: Bernoulli and binomial outcomes</a></li>
  <li><a href="#poisson-regression" id="toc-poisson-regression" class="nav-link" data-scroll-target="#poisson-regression">Poisson regression</a></li>
  <li><a href="#negative-binomial-regression" id="toc-negative-binomial-regression" class="nav-link" data-scroll-target="#negative-binomial-regression">Negative binomial regression</a></li>
  <li><a href="#gamma-regression" id="toc-gamma-regression" class="nav-link" data-scroll-target="#gamma-regression">Gamma regression</a></li>
  <li><a href="#beta-regression" id="toc-beta-regression" class="nav-link" data-scroll-target="#beta-regression">Beta regression</a></li>
  <li><a href="#non-linear-options" id="toc-non-linear-options" class="nav-link" data-scroll-target="#non-linear-options">Non-linear options</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#broken-assumption-data-are-not-independent" id="toc-broken-assumption-data-are-not-independent" class="nav-link" data-scroll-target="#broken-assumption-data-are-not-independent">Broken assumption: Data are not independent</a>
  <ul>
  <li><a href="#in-respect-to-predictors" id="toc-in-respect-to-predictors" class="nav-link" data-scroll-target="#in-respect-to-predictors">In respect to predictors</a>
  <ul>
  <li><a href="#check-for-issues" id="toc-check-for-issues" class="nav-link" data-scroll-target="#check-for-issues">Check for issues</a></li>
  <li><a href="#multivariate-methods" id="toc-multivariate-methods" class="nav-link" data-scroll-target="#multivariate-methods">Multivariate methods</a></li>
  </ul></li>
  <li><a href="#in-respect-to-measured-outcomes" id="toc-in-respect-to-measured-outcomes" class="nav-link" data-scroll-target="#in-respect-to-measured-outcomes">In respect to measured outcomes</a>
  <ul>
  <li><a href="#include-random-effects" id="toc-include-random-effects" class="nav-link" data-scroll-target="#include-random-effects">Include random effects</a>
  <ul class="collapse">
  <li><a href="#find-average-for-each-unit" id="toc-find-average-for-each-unit" class="nav-link" data-scroll-target="#find-average-for-each-unit">Find average for each unit</a></li>
  <li><a href="#blocking" id="toc-blocking" class="nav-link" data-scroll-target="#blocking">Blocking</a></li>
  <li><a href="#random-effects" id="toc-random-effects" class="nav-link" data-scroll-target="#random-effects">Random effects</a></li>
  <li><a href="#more-random-effects" id="toc-more-random-effects" class="nav-link" data-scroll-target="#more-random-effects">More random effects</a></li>
  </ul></li>
  <li><a href="#consider-a-correlation-structure-the-bigger-picture" id="toc-consider-a-correlation-structure-the-bigger-picture" class="nav-link" data-scroll-target="#consider-a-correlation-structure-the-bigger-picture">Consider a correlation structure: The bigger picture</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#errors-are-not-equal-among-groups" id="toc-errors-are-not-equal-among-groups" class="nav-link" data-scroll-target="#errors-are-not-equal-among-groups">Errors are not equal among groups</a></li>
  <li><a href="#residuals-are-not-normally-distributed" id="toc-residuals-are-not-normally-distributed" class="nav-link" data-scroll-target="#residuals-are-not-normally-distributed">Residuals are not normally distributed</a></li>
  <li><a href="#next-steps-combining-these-approaches" id="toc-next-steps-combining-these-approaches" class="nav-link" data-scroll-target="#next-steps-combining-these-approaches">Next steps: Combining these approaches</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/jsgosnell/cuny_biostats_book/edit/main/content/chapters/Linear_model_extensions.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/jsgosnell/cuny_biostats_book/blob/main/content/chapters/Linear_model_extensions.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/jsgosnell/cuny_biostats_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../content/chapters/Introduction.html">Chapters</a></li><li class="breadcrumb-item"><a href="../../content/chapters/Linear_model_extensions.html">Extending the linear model</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Linear model extensions</h1>
<p class="subtitle lead">When to trust it, and how to fix it when it’s broken</p>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- COMMENT NOT SHOW IN ANY OUTPUT: Code chunk below sets overall defaults for .qmd file; these inlcude showing output by default and looking for files relative to .Rpoj file, not .qmd file, which makes putting filesin different folders easier  -->
<p>The past several chapters/lessons have focused on linear models. Here we explore options for analysis when linear model assumptions are not met. In doing so, we review several different approaches, many of which could be (and are) the focal topic of courses and books. The goal here is to</p>
<ul>
<li><p>help you figure out when you may need to employ these methods by having you understand what they do/fix</p></li>
<li><p>give you a start on doing these analyses in R</p></li>
</ul>
<p>Citations to relevant articles papers and books are noted throughout for further exploration.</p>
<section id="sticking-with-the-linear-model" class="level2">
<h2 class="anchored" data-anchor-id="sticking-with-the-linear-model">Sticking with the linear model</h2>
<p>Linear models are useful for a number of reasons. They are a great way to unify most/many tests from classical statistics. In fact, most of the ranked tests we’ve developed can actually be run as linear models when n &gt;15. For example, we can go back to our Wilcox-Mann Whitney U tests (for 2 populations) and Kruskal-Wallis (for 3+) from the <a href="../../content/chapters/Compare_means_among_populations.html">comparing means among groups chapter</a> and note the outcome from a <em>wilcox.test</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>two_species_subset <span class="ot">&lt;-</span> iris[iris<span class="sc">$</span>Species<span class="sc">!=</span><span class="st">"setosa"</span>,]</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wilcox.test</span>(Sepal.Length <span class="sc">~</span> Species, two_species_subset)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  Sepal.Length by Species
W = 526, p-value = 5.869e-07
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<p>is very close to a linear model predicting the signed rank of the data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(car)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>signed_rank <span class="ot">=</span> <span class="cf">function</span>(x) <span class="fu">sign</span>(x) <span class="sc">*</span> <span class="fu">rank</span>(<span class="fu">abs</span>(x))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(<span class="fu">lm</span>(<span class="fu">signed_rank</span>(Sepal.Length) <span class="sc">~</span> Species, two_species_subset), <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: signed_rank(Sepal.Length)
            Sum Sq Df F value    Pr(&gt;F)    
(Intercept)  64872  1 102.378 &lt; 2.2e-16 ***
Species      20967  1  33.089 1.003e-07 ***
Residuals    62098 98                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>In fact, we could run simulations and show that p values from these 2 approaches are highly correlated (now you know what that means) with a <span class="math inline">\(\beta\)</span> of almost 1 (from <span class="citation" data-cites="lindelov">Lindelov (<a href="#ref-lindelov" role="doc-biblioref">n.d.</a>)</span>).</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-3-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Linear models are also extremely robust. Consider the basic assumptions of a linear model</p>
<p><span class="math display">\[
\epsilon \approx i.i.d.\ N(\mu,\sigma)
\]</span></p>
<p>Although the residuals are meant to be homoscedastic (equal or constant across all groups), it turns out the model is robust of when the largest group variance is 4-10x larger than the smallest group variance and sample sizes are approximately equal <span class="citation" data-cites="blanca2018 fox2015 zuur2010">(<a href="#ref-blanca2018" role="doc-biblioref">Blanca et al. 2018</a>; <a href="#ref-fox2015" role="doc-biblioref">Fox 2015</a>; <a href="#ref-zuur2010" role="doc-biblioref">A. F. Zuur, Ieno, and Elphick 2010</a>)</span>, though highly uneven group sizes begin to cause issues <span class="citation" data-cites="blanca2018">(<a href="#ref-blanca2018" role="doc-biblioref">Blanca et al. 2018</a>)</span>.</p>
<p>Similarly, non-normal data is not an issue in an of itself <span class="citation" data-cites="knief2021">(<a href="#ref-knief2021" role="doc-biblioref">Knief and Forstmeier 2021</a>)</span>. This is partially because the assumptions are focused on residuals, but also because the procedure is highly robust <span class="citation" data-cites="blanca2017">(<a href="#ref-blanca2017" role="doc-biblioref">Blanca et al. 2017</a>)</span>. This finding further supports the graphical consideration of assumptions , especially since many tests for normality are conservative (samples are almost never <em>perfectly</em> normal, and slight deviations are easier to pick up with larger sample sizes despite the fact the central limit theorem suggests this is when the issue is least important) <span class="citation" data-cites="zuur2010 shatz2023">(<a href="#ref-zuur2010" role="doc-biblioref">A. F. Zuur, Ieno, and Elphick 2010</a>; <a href="#ref-shatz2023" role="doc-biblioref">Shatz 2023</a>)</span>.</p>
<details>
<summary>
How can residuals be normal when data is not?
</summary>
<p>This is a common question. Sometimes the outcome variable is not normally distributed due to the impact of various factors. Consider this example of sparrow weight from <span class="citation" data-cites="zuur2010">A. F. Zuur, Ieno, and Elphick (<a href="#ref-zuur2010" role="doc-biblioref">2010</a>)</span>. The raw data may look like this…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>Sparrows <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file=</span><span class="st">"data/SparrowsElphick.txt"</span>, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>Sparrows<span class="sc">$</span>fMonth<span class="ot">&lt;-</span><span class="fu">factor</span>(Sparrows<span class="sc">$</span>Month,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">levels =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">7</span>, <span class="dv">8</span>, <span class="dv">9</span>, <span class="dv">10</span>),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"May"</span>, <span class="st">"June"</span>, <span class="st">"July"</span>, <span class="st">"August"</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="st">"Sept."</span>, <span class="st">"Oct."</span>))</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>Sparrows<span class="sc">$</span>I1 <span class="ot">&lt;-</span> Sparrows<span class="sc">$</span>fMonth <span class="sc">==</span><span class="st">"June"</span> <span class="sc">|</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>               Sparrows<span class="sc">$</span>fMonth <span class="sc">==</span><span class="st">"July"</span> <span class="sc">|</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>               Sparrows<span class="sc">$</span>fMonth <span class="sc">==</span><span class="st">"August"</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Sparrows[Sparrows<span class="sc">$</span>I1<span class="sc">==</span> T, ], <span class="fu">aes</span>(<span class="at">x=</span>wt))<span class="sc">+</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()<span class="sc">+</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>( <span class="at">x =</span> <span class="st">"Weight (g)"</span>, <span class="at">main =</span> <span class="st">"Weight of sparrows in study"</span>, <span class="at">y =</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>which is skewed (albeit slightly). However, if we consider the impact of month…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Sparrows[Sparrows<span class="sc">$</span>I1<span class="sc">==</span> T, ], <span class="fu">aes</span>(<span class="at">x=</span>wt, <span class="at">fill=</span>fMonth))<span class="sc">+</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()<span class="sc">+</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>( <span class="at">x =</span> <span class="st">"Weight (g)"</span>, <span class="at">main =</span> <span class="st">"Weight of sparrows in study"</span>, <span class="at">y =</span> <span class="st">"Frequency"</span>, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">fill=</span> <span class="st">"Month"</span>)<span class="sc">+</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_col</span>(<span class="sc">~</span>fMonth)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We would find the residuals to be much better behaved.</p>
<p>Similarly, highly skewed data may arise due to unequal sample sizes (which may pose their own problems, but not outright), but models using this data may meet assumptions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">8</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>example_skewed <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Population=</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"a"</span>,<span class="dv">40</span>),</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                                           <span class="fu">rep</span>(<span class="st">"b"</span>, <span class="dv">30</span>),</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                                           <span class="fu">rep</span>(letters[<span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">each=</span><span class="dv">15</span>)), </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>                             <span class="at">Growth =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">40</span>,<span class="dv">1</span>),</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">rnorm</span>(<span class="dv">30</span>, <span class="dv">4</span>),</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">rnorm</span>(<span class="dv">15</span>, <span class="dv">7</span>),</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">rnorm</span>(<span class="dv">15</span>,<span class="dv">10</span>)))</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(example_skewed,<span class="fu">aes</span>(<span class="at">x=</span>Growth))<span class="sc">+</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()<span class="sc">+</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Some value"</span>,</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="st">"Example of skewed data appearing due to unequal sample size"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(example_skewed,<span class="fu">aes</span>(<span class="at">x=</span>Growth, <span class="at">fill=</span>Population))<span class="sc">+</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>()<span class="sc">+</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"Some value"</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="st">"Example of skewed data appearing due to unequal sample size"</span>)<span class="sc">+</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_col</span>(<span class="sc">~</span>Population)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">lm</span>(Growth<span class="sc">~</span>Population, example_skewed))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-6-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-6-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-6-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-6-6.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(<span class="fu">lm</span>(Growth<span class="sc">~</span>Population, example_skewed), <span class="at">type=</span><span class="st">"III"</span>)                              </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: Growth
             Sum Sq Df F value    Pr(&gt;F)    
(Intercept)   38.75  1  32.436 1.344e-07 ***
Population  1007.71  3 281.175 &lt; 2.2e-16 ***
Residuals    114.69 96                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</details>
<p>These issues have led some authors to argue violations of the linear model assumptions are less dangerous than trying new, and often less-tested, techniques that may inflate type I error rates <span class="citation" data-cites="knief2021 warton2016">(<a href="#ref-knief2021" role="doc-biblioref">Knief and Forstmeier 2021</a>; <a href="#ref-warton2016" role="doc-biblioref">D. I. Warton et al. 2016</a>)</span>. However, new techniques (or sometimes old techniques that are new to a given field) may be more appropriate when assumptions are clearly broken <span class="citation" data-cites="warton2010 reitan2016 geissinger2022">(<a href="#ref-warton2010" role="doc-biblioref">D. Warton and Hui 2010</a>; <a href="#ref-reitan2016" role="doc-biblioref">Reitan and Nielsen 2016</a>; <a href="#ref-geissinger2022" role="doc-biblioref">Geissinger et al. 2022</a>)</span>. In this section we explore common-ish approaches for analyzing data when the assumptions of a linear model are broken. Our goal here is to introduce these methods. Each could be the focus of their own class, book, or much larger study. Fortunately most can be viewed as extensions to our existing knowledge, although many of the assumptions and techniques for them are less developed/still being argued about.</p>
<p>The various assumptions related to linear models may be prioritized on their relative importance. One such order is provided (roughly) by <span class="citation" data-cites="gelman2006">Gelman and Hill (<a href="#ref-gelman2006" role="doc-biblioref">2006</a>)</span>.</p>
<ul>
<li><p>Model validity</p>
<ul>
<li>As noted in the multiple regression chapter, we only should investigate relationships we have a mechanism to explain</li>
</ul></li>
<li><p>Linear relationship and additive effects of predictor variables</p></li>
<li><p>Errors are</p>
<ul>
<li><p>independently distributed</p></li>
<li><p>identical (homoscedastic)</p></li>
<li><p>follow a normal distribution</p></li>
</ul></li>
</ul>
<p>Many datasets will violate multiple of these assumptions simultaneously, so addressing issues is often best resolved by understanding <em>why</em> this is happening.</p>
</section>
<section id="broken-assumption-linear-relationship-is-inappropriate" class="level2">
<h2 class="anchored" data-anchor-id="broken-assumption-linear-relationship-is-inappropriate">Broken assumption: Linear relationship is inappropriate</h2>
<p>A central (and often overlooked assumption) of linear models is that the relationship between the predictors (remember, not the actual data!) and the outcome variable is linear and additive. When the relationship is not linear, the resulting residuals are often not appropriately distributed as well.</p>
<details>
<summary>
What’s linear anyway?
</summary>
<p>To be clear, the linear model only focuses on the linear and additive relationship between the predictors and the outcome variable (this will become more important/obvious later in this section!). The model doesn’t <em>know</em> what the variables are. For that reason, we can add predictor variables to a model that are squares or cubes of predictors, or we can transform the outcome variable. We just need the <span class="math inline">\(Y = X\beta\)</span> relationship to be additive and linear.</p>
</details>
<p>When non-linearity occurs, several options to address it exist. The best approach may depend on why the relationship is non-linear, as relationships among variables may be non-linear for a number of reasons. For example, the outcome may not actually be continuous (e.g.&nbsp;counts. proportions), and thus true linear relationships are not possible; outcomes may also be related to the predictors in different ways (e.g., logarithmic).</p>
<section id="transform-the-data-least-advisable-sometimes" class="level3">
<h3 class="anchored" data-anchor-id="transform-the-data-least-advisable-sometimes">Transform the data (least advisable, sometimes)</h3>
<p>One way to address non-linearity is to transform the data (typically focusing on the dependent variable) so that the resulting variable meets the linear model assumptions (and thus uses the strengths of linear models that we noted above). As shown above, our rank-based approaches are using a similar method (not technically the same, but it works for larger sample sizes). This approach was often used in the past <span class="citation" data-cites="warton2010">(e.g., arc-sin transforms of proportion data, <a href="#ref-warton2010" role="doc-biblioref">D. Warton and Hui 2010</a>)</span> and supported by various approaches. For example, the Box-Cox transformation helped researchers find the best way to transform data to reduce issues with the distribution of residuals; this method also tended to impact linearity and differences in variances.</p>
<p>Two things should be noted regarding this approach and transformations in general:</p>
<ul>
<li><p>The Box-Cox approach requires a model - it still focused on transforming data so that <em>residuals</em> meet assumptions. Data should not be transformed before model assumptions are analyzed to ensure it is necessary.</p></li>
<li><p>Similarly (and already noted!). non-normal data may lead to models with normal residuals. However, normal data (or data transformed to be <em>more</em> normal) does typically lead to normal residuals, so if residuals are not normal transformations may help.</p></li>
<li><p>The transformed variable is <em>now</em> linear in respect to the predictors. This highlights the actual assumption of the model. Similarly, higher-terms (squares, cubes, etc) may be added to a linear model. The model does not care what the data represent - it only focuses on if linear relationships exist among them.</p></li>
<li><p>Transformations can make model interpretation and prediction difficult.</p></li>
</ul>
<p>If the decision is made to transform the data, several approaches exist. Some are driven by the distribution of the data, and all depend on it. For example, log and related root transformations are useful for right-skewed data, but some can only be carried out for non-negative (e.g., square root) or strictly positive (e.g., log) values. To address this for log transformations, a small value is often added to 0 measurements.</p>
<p>Let’s use data (not residuals) to show how different distributions of data connect to q-q plots and consider possible fixes (always fit a model first for real analysis!). For example, we can return to our right-skewed blue jay data <a href="../../content/chapters/summarizing_data.html">from the summarizing data chapter</a>(idea from <span class="citation" data-cites="hunt">(<a href="#ref-hunt" role="doc-biblioref">Hunt, n.d.</a>)</span>) .</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>blue_jays <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">rbeta</span>(<span class="dv">10000</span>,<span class="dv">2</span>,<span class="dv">8</span>)<span class="sc">*</span><span class="dv">100</span><span class="sc">+</span><span class="fu">runif</span>(<span class="dv">10000</span>,<span class="dv">60</span>,<span class="dv">80</span>),<span class="dv">3</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(blue_jays), </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x=</span>blue_jays)) <span class="sc">+</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>( <span class="at">fill=</span><span class="st">"blue"</span>, <span class="at">color=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Weight of Westchester blue jays"</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">x=</span> <span class="st">"Weight (g)"</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Note the right-skewed data shows a convex curve on the Q-Q plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(blue_jays)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(blue_jays)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>To help you understand Q-Q plots, remember they are comparing the relative spread of data (quantiles) from a target and theoretical distribution. For right-skewed data, points are shifted right (both smallest and largest observations are larger than you would expect from a normal distribution).</p>
<p>Conversely, we could also return to our left-skewed cardinal data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>cardinals <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">rbeta</span>(<span class="dv">10000</span>,<span class="dv">10</span>,<span class="dv">2</span>)<span class="sc">*</span><span class="dv">50</span><span class="sc">+</span><span class="fu">runif</span>(<span class="dv">10000</span>,<span class="dv">5</span>,<span class="dv">10</span>),<span class="dv">3</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(cardinals), </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x=</span>cardinals)) <span class="sc">+</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>( <span class="at">fill=</span><span class="st">"red"</span>, <span class="at">color=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Weight of Westchester cardinals"</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">x=</span> <span class="st">"Weight (g)"</span>,</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>and note a concave shape is seen in the Q-Q plot,as all points are shifted left.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(cardinals)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(cardinals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>For this type of data, power transformations (raising the variable to the 2, 3, or higher power) may be useful.</p>
<p>Meanwhile, our uniformly-distributed distributed robin data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>rochester <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">c</span>(<span class="fu">runif</span>(<span class="dv">1000</span>,<span class="dv">75</span>,<span class="dv">85</span>)),<span class="dv">3</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(rochester), </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x=</span>rochester)) <span class="sc">+</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>( <span class="at">fill=</span><span class="st">"pink"</span>, <span class="at">color=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Weight of Rochester robins"</span>,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">x=</span> <span class="st">"Weight (g)"</span>,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>shows as a s-shape on the Q-Q plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(rochester)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(rochester)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>because it is <em>under-dispersed</em> (has no tails, and thus the lower points are larger than expected and the higher points are smaller than expected). Alternatively, data may be <em>over-dispersed</em>, like this (fake) finch data where lower points are lower than expected and higher points are higher.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(VGAM)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>finch <span class="ot">&lt;-</span> <span class="fu">rlaplace</span>(<span class="dv">1000</span>,<span class="at">location=</span><span class="dv">50</span>, <span class="at">scale=</span><span class="dv">4</span>)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(finch), </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x=</span>finch)) <span class="sc">+</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>( <span class="at">fill=</span><span class="st">"cyan"</span>, <span class="at">color=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Weight of finches"</span>,</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">x=</span> <span class="st">"Weight (g)"</span>,</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(finch)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(finch)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Over- and under-dispersed data may mean there’s a missing factor in your analysis. For example, our bi-modal woodpecker data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>woodpeckers <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">60</span>,<span class="dv">4</span>),<span class="fu">rnorm</span>(<span class="dv">100</span>,<span class="dv">80</span>,<span class="dv">4</span>)),<span class="dv">3</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(woodpeckers), </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x=</span>woodpeckers)) <span class="sc">+</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>( <span class="at">fill=</span><span class="st">"orange"</span>, <span class="at">color=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Weight of  Westchester woodpeckers"</span>,</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">x=</span> <span class="st">"Weight (g)"</span>,</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>is under-dispersed due the shape of the distribution, but there also signs of other issues.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">qqnorm</span>(woodpeckers)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">qqline</span>(woodpeckers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Under-dispersion could also happen if data are bounded (e.g., by practicality or due to a measurement issue). Over-dispersion can similarly occur if the model does not account for variability (e.g., missing factors, non-linear relationship) and/or outliers, which might be related to the underlying form of the data <span class="citation" data-cites="payne2017">(<a href="#ref-payne2017" role="doc-biblioref">Payne et al. 2017</a>)</span> (more to come on this). In this way, over- and under-dispersion relate to both the linear relationship.</p>
</section>
<section id="different-data-require-a-different-approach-introducing-generalized-linear-models" class="level3">
<h3 class="anchored" data-anchor-id="different-data-require-a-different-approach-introducing-generalized-linear-models">Different data require a different approach: Introducing generalized linear models</h3>
<p>Note: much of this follows/borrows heavily from the <a href="https://r.qcbs.ca/workshop06/book-en/index.html">QCBS workshop on generalized linear models</a><span class="citation" data-cites="series">(<a href="#ref-series" role="doc-biblioref">Series and QCBS R. Workshop, n.d.</a>)</span>.</p>
<p>The linear relationship may be inappropriate because our data doesn’t fit it! For example, if we are modeling proportions, estimates less than 0 or below 1 do not make sense, but a linear model doesn’t account for that. The same issues arise for binary outcomes (outcome must be 0 or 1) and count data (where outcomes can’t be non-integer). Along with impacts on the assumption of linearity, data like this may also lead to unequal variances and lack of normality for residuals. While we may be able to transform the response variable to make relationships linear, another option is to translate the <em>link</em> function.</p>
<p>Although we haven’t fully explained it before, a linear model contains three components. There is always a random component that focuses on the distribution of the data. This typically requires a few parameters (or at least one!). For example, for normally-distributed data, the two parameters that describe the distribution are the mean and the variance. Note these parameters are independent (we covered all this when we <a href="../../content/chapters/Tests_for_continuous_data_from_one_sample.html">introduced normality</a>. There is also a systematic component, where a number of covariates relate linearly to a predictor. Finally, there is the link function, which connects that predictor to the data distribution (this maintains the linearity of the model). Notice this means the link connects to the distribution of the data, not the data itself (which is what transforming the data is doing).</p>
<p>So far we have focused on the <em>mean</em> of the data, and the estimate is the mean. For example, in simple linear regresssion we predict the mean response of <span class="math inline">\(Y\)</span> for each level of <span class="math inline">\(X\)</span>. This is equivalent to predicting <span class="math inline">\(\mu\)</span>, so we can write</p>
<p><span class="math inline">\(Y\)</span>= <span class="math inline">\(X\beta\)</span></p>
<p>or</p>
<p><span class="math inline">\(\mu\)</span>= <span class="math inline">\(X\beta\)</span></p>
<p>where:</p>
<ul>
<li><p><span class="math inline">\(Y\)</span> (or <span class="math inline">\(\mu\)</span> is the predicted value of the response variable</p></li>
<li><p><span class="math inline">\(X\)</span> is the model matrix (<em>i.e.</em> predictor variables)</p></li>
<li><p><span class="math inline">\(β\)</span> is the estimated parameters from the data (<em>i.e.</em> intercept and slope). <span class="math inline">\(XB\)</span> is called the <strong>linear predictor</strong></p></li>
</ul>
<p>Earlier we also had <span class="math inline">\(E\)</span> , an error term that account for deviations in our data from the expected outcome. We still have error but will not write that for clarity at this point.</p>
<p>Since <span class="math inline">\(Y = \mu\)</span>, the link has been implied, so we can think of this as</p>
<p><span class="math display">\[\mu = \underbrace{X\beta}_{Linear~component}\]</span></p>
<p>owever, we can <em>generalize</em> this setup to include other distributions in the exponential family (and now others <span class="citation" data-cites="applied2022">(<a href="#ref-applied2022" role="doc-biblioref"><span>“Applied Regression Analysis and Generalized Linear Models”</span> 2022</a>, ch.&nbsp;15)</span>). The exponential family is a group of distributions where the probability distribution follows a specific formula. While we won’t develop all the math, exponential family distributions all have a canonical parameter (the mean for Gaussian, or normal, data) and a dispersion parameter (the variance for normal data). In doing that, we predict</p>
<p><span class="math display">\[\eta_i = \underbrace{X\beta}_{Linear~component}\]</span></p>
<p>and link <span class="math inline">\(\eta\)</span> to our true parameter of interest using a link function.</p>
<p><span class="math display">\[g(\mu) = \eta\]</span></p>
<p>The main idea is we can keep a systematic component with a linear relationship for data that doesn’t really have that relationship and that don’t fit the normal approximation. We call models using this approach <em>generalized linear models</em> (which are different than general linear models, which is another term used to describe our “regular” linear models). These models are also known as GLM or GLIM (we’ll use the GLM jargon here).</p>
<p>For example, we can consider binomial data where <span class="math inline">\(Y ∼ B(n, p)\)</span>. This type of data has to be between 0 and 1 (and only takes fractional values in between that depend on the sample size!). We introduced this type of data in the <a href="../../content/chapters/Binomial.html">chapter introducing binomial data</a>. If we use a basic linear model, we would predict outcomes that don’t match the data. However, instead of focusing on the <span class="math inline">\(p\)</span> parameter like we did before, let’s consider the odds of something happening. We can estimate</p>
<p><span class="math display">\[
\mu = N*p
\]</span></p>
<p>(we introduced this normal approximation earlier too), and find the odds as</p>
<p><span class="math display">\[\text{odds}_i = \frac{\mu_i}{1-\mu_i}\]</span></p>
<p>this takes us to a <code>0</code> to <code>+Inf</code> scale for our predictor. We can then take the log of this outcome, which we call the logit</p>
<p><span class="math display">\[\text{logit}(\mu_i) = \log(\frac{\mu_i}{1-\mu_i})\]</span></p>
<p>which can now takes values from <code>-Inf</code> to <code>+Inf</code> . If we predict this value, so that</p>
<p><span class="math display">\[g(\mu) = \eta = \text{logit}(\mu_i) = \log(\frac{\mu_i}{1-\mu_i}) = X\beta \]</span></p>
<p>we have constructed a <em>generalized linear model</em>.</p>
<p>For this class I don’t expect you to derive the appropriate links. Common links for various types of data are shown below, and in R we can identify these using a <em>family</em> argument in the <em>glm</em> function (introduced below).</p>
<table class="table">
<colgroup>
<col style="width: 21%">
<col style="width: 20%">
<col style="width: 19%">
<col style="width: 19%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th>Distribution of <span class="math inline">\(Y\)</span></th>
<th>Link function name</th>
<th>Link function</th>
<th>Model</th>
<th><code>R family</code></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Normal</td>
<td>Identity</td>
<td><span class="math inline">\(g(\mu) = \mu\)</span></td>
<td><span class="math inline">\(\mu = \mathbf{X} \boldsymbol{\beta}\)</span></td>
<td><code>gaussian(or just run lm)</code></td>
</tr>
<tr class="even">
<td>Bernoulli/Binomial</td>
<td>Logit</td>
<td><span class="math inline">\(g(\mu) = \log\left(\dfrac{\mu}{1-\mu}\right)\)</span></td>
<td><span class="math inline">\(\log\left(\dfrac{\mu}{1-\mu}\right) = \mathbf{X} \boldsymbol{\beta}\)</span></td>
<td><code>binomial(default link is logit)</code></td>
</tr>
<tr class="odd">
<td>Poisson</td>
<td>Log</td>
<td><span class="math inline">\(g(\mu) = \log(\mu)\)</span></td>
<td><span class="math inline">\(-\mu^{-1} = \mathbf{X} \boldsymbol{\beta}\)</span></td>
<td><code>poisson</code></td>
</tr>
<tr class="even">
<td>Exponential</td>
<td>Negative Inverse</td>
<td><span class="math inline">\(g(\mu) = -\mu^{-1}\)</span></td>
<td><span class="math inline">\(\log(\mu) = \mathbf{X} \boldsymbol{\beta}\)</span></td>
<td><code>Gamma</code></td>
</tr>
</tbody>
</table>
<p>However, the major issue is realizing these links mean interpreting coefficient values requires an additional step to <em>transform</em> them back to impacts on the actual parameter of interest. Examples are provided below.</p>
<p>When should we use GLMs? These various data types (and some other distributions we’ll demonstrate) match some common real-world scenarios. <span class="citation" data-cites="bolker2008">Bolker (<a href="#ref-bolker2008" role="doc-biblioref">2008</a>)</span> has an excellent review of when various distributions should be used. The table below also lists some of the more common distributions and notes on their usage.</p>
<table class="table">
<caption>Connections between data distributions and real-world data</caption>
<colgroup>
<col style="width: 25%">
<col style="width: 74%">
</colgroup>
<thead>
<tr class="header">
<th>Distribution</th>
<th>Typical usage, notes, and examples</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong><em>Continuous</em></strong></td>
<td></td>
</tr>
<tr class="even">
<td>Gaussian/normal</td>
<td>continuous variables, or variables connected to large sample sizes and not extreme values (remember the z-test approximation for <span class="math inline">\(\chi^2\)</span> tests?)</td>
</tr>
<tr class="odd">
<td>Gamma</td>
<td>Amount of time until a certain number of events takes place (how long do you wait for a parking spot?); variance changes with mean; related to negative binomial</td>
</tr>
<tr class="even">
<td>Exponential</td>
<td>amount of time for a single event to occur; related to geometric</td>
</tr>
<tr class="odd">
<td>Beta</td>
<td>distribution of success parameter in binomial trials with a-1 success and b-1 failures; bounded by 0 -1 (not inclusive)</td>
</tr>
<tr class="even">
<td><em>Discrete</em></td>
<td></td>
</tr>
<tr class="odd">
<td>Bernoulli/binomial</td>
<td>data that can be divided into successes and failures</td>
</tr>
<tr class="even">
<td>Poisson</td>
<td>Fits count data where all events are independent (random); Count data where low counts and 0’s are common</td>
</tr>
<tr class="odd">
<td>Negative binomial</td>
<td>Number of failures before a set number of successes occur; Poisson-ish where variance is larger than the mean (good for overdispersion!) or where the mean (and thus variance) varies</td>
</tr>
<tr class="even">
<td>Geometric</td>
<td>Number of trials until you get a single failure</td>
</tr>
<tr class="odd">
<td><em>Combination</em></td>
<td></td>
</tr>
<tr class="even">
<td>Beta binomial</td>
<td>Binomial-ish but the success parameter (and thus variance) varies across trials</td>
</tr>
</tbody>
</table>
<p>Different distributions also have different relationships between the variance and mean (or between the canonical and dispersion parameter for distributions in the exponential family). For normal data, there is <em>no</em> relationship, so the variance is constant. This is why we have assumed <em>homoscedasticity</em> of residuals for linear models. This is not true for other families; for example, the variance in binomial data depends on the <em>p</em> parameter (as we noted in the <a href="../../content/chapters/Binomial.html">chapter introducing binomial data</a>), and for the Poisson distribution (used for count data) the variance is the mean.</p>
<p>These relationships are reflected in the assumptions for a GLM (though everyone does not agree on what they are!). First, we need to ensure the correct family (and thus mean-variance link) is chosen.While we still need independence of the residuals (or some extension, see below), they no longer need to be normally distributed <span class="citation" data-cites="zuur2007 zuur2009 fox2016">(<a href="#ref-zuur2007" role="doc-biblioref">A. Zuur, Ieno, and Smith 2007, 86–87</a>; <a href="#ref-zuur2009" role="doc-biblioref">A. Zuur et al. 2009, sec. 9.8.3</a>; <a href="#ref-fox2016" role="doc-biblioref">Fox 2016</a>)</span> (though normality may hold at large sample sizes for Pearson <span class="citation" data-cites="agresti2007">(<a href="#ref-agresti2007" role="doc-biblioref">Agresti 2007, 87</a>)</span> or deviance <span class="citation" data-cites="montgomery2012">(<a href="#ref-montgomery2012" role="doc-biblioref">Montgomery, Peck, and Vining 2012, sec. 13.4.4</a>)</span> <!-- https://stats.stackexchange.com/questions/92394/checking-residuals-for-normality-in-generalised-linear-models; https://online.stat.psu.edu/stat504/lesson/6/6.1 --> residuals, and some authors argue normality is required for testing <span class="citation" data-cites="feng2020">(<a href="#ref-feng2020" role="doc-biblioref">Feng, Li, and Sadeghpour 2020</a>)</span>) .</p>
<p>In general, plotting the residuals against <span class="citation" data-cites="zuur2009">(<a href="#ref-zuur2009" role="doc-biblioref">A. Zuur et al. 2009</a>)</span></p>
<ul>
<li><p>the predicted values</p>
<ul>
<li>to check for model fit</li>
</ul></li>
<li><p>each explanatory variable in model</p>
<ul>
<li>to check for linearity</li>
</ul></li>
<li><p>each explanatory not retained/used in model</p>
<ul>
<li>to check for missing patterns</li>
</ul></li>
<li><p>against time and space (if applicable)</p>
<ul>
<li>to check for patterns and correlation</li>
</ul></li>
</ul>
<p>will be useful. Any patterns may suggest missing predictors or lack of independence/correlation among measurements. Patterns in the residuals may also indicate the incorrect family has been chosen <span class="citation" data-cites="zuur2009">(<a href="#ref-zuur2009" role="doc-biblioref">A. Zuur et al. 2009, sec. 9.8.4</a>)</span>; for families with fixed dispersion parameters, inspection of the estimated dispersion parameter can also be a related diagnostic.</p>
<details>
<summary>
What are residuals for a GLM?
</summary>
<p>Residuals for a Gaussian/normal GLM (what we’ve been doing) are easy to calculate. We simply subtract the estimate for each data point (which is the average for similar observations!) from the observed. However, though these <em>raw</em> or <em>response</em> residuals exist for GLM, they rarely make sense. Variance may increase with mean for some distributions, for example, or outcomes may be binary (so every residual is 0 or 1!).</p>
<p>To account for this, we could divide the residuals by the standard deviation of the observations. This normalizes the residuals and leads to <em>Pearson residuals</em>. Another option is based on something called deviance. Deviance is similar to sums of squares but based on likelihood (which we shortly see is what we use to test for significance in GLM); it can also be considered the generalized form of variance or residual sums of squares. It is calculated as the difference of log-likelihoods between the focal model and the saturated model (or a model that has a coefficient for every observation and thus will have perfect fit). Deviance residuals consider how important each point is to the overall deviance. You can also compare the deviance of a model to a perfect fit (no deviance/saturated) model and an intercept-only model (worst fit, D<sub>O</sub>) and generate a pseudo-R<sup>2</sup> measure using the formula</p>
<p><span class="math display">\[
1-\frac{D}{D_O}
\]</span></p>
<p>This is known as McFadden’s pseudo-R<sup>2</sup> value.</p>
<p>Another option are quantile residuals. These are the focus of the DHARMa package and should be uniformly distributed for appropriate models.</p>
</details>
<p>Once developed, GLMs can use most of the tools we developed for basic linear models. However, a few key differences exist. Estimation of parameters and significance for these models rely on maximum-likelihood methods. It should be noted that solving for coefficients also requires iterative approaches due to non-linearity; common methods (not described here) include the Newton-Raphson or Fisher scoring methods. The key point here for practice is any noted issues with optimization are due to this process. The ratio of likelihood values from various nested tests follow a <span class="math inline">\(\chi^2\)</span> distribution and can thus provide a p-value to compare possible models as noted in <a href="../../content/chapters/Compare_proportions_among_populations.html">introducing G tests</a>.This may be represented as a Z test for single parameters (since a <span class="math inline">\(\chi^2\)</span> distribution is a Z distribution squared - sometimes labelled a Wald test/statistic). F test reappear for distributions where we estimate the dispersion parameter (like the normal), but otherwise we can use the dispersion parameter to check that the right approach was taken.</p>
<p>As you may infer from the numerous distributions mentioned above, there are a variety of GLMs. Here we outline some of the more common ones. Each of these deserves a much fuller treatment - remember, the intent here is to get you to a place to where you can begin working with them. <span class="citation" data-cites="harrison2018">Harrison et al. (<a href="#ref-harrison2018" role="doc-biblioref">2018</a>)</span> (which also covers mixed models (coming up!)) provides an excellent review of these models, and fuller treatments are provided by <span class="citation" data-cites="zuur2009">A. Zuur et al. (<a href="#ref-zuur2009" role="doc-biblioref">2009</a>)</span> (chapters 8-11) and <span class="citation" data-cites="fox2016">Fox (<a href="#ref-fox2016" role="doc-biblioref">2016</a>)</span> (chapters 14 - 15). For each I provide the most relevant math, an example (or two), and some immediate extensions.</p>
<section id="logistic-regression-bernoulli-and-binomial-outcomes" class="level4">
<h4 class="anchored" data-anchor-id="logistic-regression-bernoulli-and-binomial-outcomes">Logistic regression: Bernoulli and binomial outcomes</h4>
<p>Logistic regression <span class="citation" data-cites="warton2010">(<a href="#ref-warton2010" role="doc-biblioref">D. Warton and Hui 2010</a>)</span> focuses on modelling yes/no or success/failure outcomes (Bernoulli or binomial data) using the <em>logit</em> link (thus logistic regression). As noted above, this type of data uses the logit link. This link also allows coefficients to be interpreted as odds and odds ratios (introduced in the <a href="../../content/chapters/Compare_proportions_among_populations.html">comparing proportions among groups</a>- we’ll continue to see connections to this chapter for logistic regression). For a given coefficient <span class="math inline">\(\beta\)</span>, the logit link also means we can say a 1 unit increase in that single variable leads to a change in the odds ratio of <span class="math inline">\(e^{\beta}\)</span>.</p>
<details>
<summary>
Some math if you want/need it
</summary>
<p>Assume we have a binary outcome <span class="math inline">\(y\)</span> and one covariates <span class="math inline">\(x_1\)</span>. For this situation, the probability of a successful outcome ( <span class="math inline">\(y = 1\)</span> ) is given by:</p>
<p><span class="math display">\[Pr(y_i = 1) = p = g^{-1}(X\beta)\]</span></p>
<p>where <span class="math inline">\(g^{-1}()\)</span> is the <strong>inverse link function</strong>.</p>
<p>For the identity link, the interpretation of the <span class="math inline">\(\beta_1\)</span> coefficient (remember,</p>
<p><span class="math display">\[
\beta = \begin{bmatrix}
\beta_0\\
\beta_1\
\end{bmatrix}
\]</span></p>
<p>is straighforward: For one-unit increase in <span class="math inline">\(x_1\)</span>, <span class="math inline">\(\beta_1\)</span> dictates a constant difference in the outcome.</p>
<p><span class="math display">\[\Delta{y_i} = (\beta_0 + \beta_1(\color{red}{x_{1i} + 1})) - (\beta_0 + \beta_1x_{1i})\]</span></p>
<p><span class="math display">\[\Delta{y_i} = \beta_1\]</span></p>
<p>In a linear logistic model w$, we have:</p>
<p><span class="math display">\[log({\dfrac{p}{1-p}})=\beta_0 + \beta_1x_{1i}\]</span></p>
<p>This corresponds to a log odds ratio! We can the use an exponential function to rewrite that model to get the odds ratio:</p>
<p><span class="math display">\[\dfrac{p}{1-p}=exp(\beta_0 + \beta_1x_{1i} + \beta_2x_{2i})\]</span></p>
<p>If we want to convert the odds into probability, given a coefficient <span class="math inline">\(\alpha\)</span> we would use the <strong>inverse logit link function</strong> (also known as the logistic function):</p>
<p><span class="math display">\[ Pr(y_i = 1) = logit^{-1}(\alpha) = \dfrac{1}{1 + exp(-\alpha)} = (\dfrac{1}{1 + exp(-\alpha)}) * (\dfrac{exp(\alpha)}{exp(\alpha)}) = \dfrac{exp(\alpha)}{exp(\alpha) + 1}\]</span></p>
<p>Now going back to our model, this gives us:</p>
<p><span class="math display">\[Pr(y_i = 1) = \dfrac{exp(\beta_0 +\beta_1 x_{1i})}{1 + exp{(\beta_0 + \beta_1x_{1i})}}\]</span></p>
<p>Since the inverse link is nonlinear, it is difficult to interpret the coefficient. However, we can look what happens to the differences for a one-unit change to <span class="math inline">\(x_1\)</span>:</p>
<p><span class="math display">\[\Delta{y_i} = \dfrac{\exp(\beta_0 + \beta_1(\color{red}{x_{1i} + 1}))}{1 + \exp{(\beta_0 + \beta_1(\color{red}{x_{1i} + 1}) })} - \dfrac{\exp(\beta_0 + \beta_1x_{1i} )}{1 + \exp{(\beta_0 + \beta_1x_{1i}  )}}\]</span></p>
<p><span class="math display">\[\Delta{y_i} = \exp(\beta_1)\ \]</span></p>
<p>As <span class="math inline">\(x_1\)</span> increases by one unit, the odds increase by a factor of <span class="math inline">\(\exp(\beta_1)\)</span>. Note that the odds values here are considered when all other parameters are kept constant.</p>
<p>From <span class="citation" data-cites="workshop"><em>Workshop 6: Generalized Linear Models</em> (<a href="#ref-workshop" role="doc-biblioref">n.d.</a>)</span></p>
</details>
<p>This also means no impact is shown as <span class="math inline">\(\beta = 0\)</span>, with positive and negative numbers implying increases or decreases in odds, which is convenient. Other link options (not discussed here) include the probit and complementary log-log approaches. For a final piece of math, note the dispersion parameter for binomial data is not estimated - it is assumed to be 1.</p>
<p>For example, <span class="citation" data-cites="blais2023">Blais, Johnson, and Koprowski (<a href="#ref-blais2023" role="doc-biblioref">2023</a>)</span> investigated how various factors impacted behavior in garter snakes (<em>Thamnophis cyrtopsis</em>).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://https://commons.wikimedia.org/wiki/File:Thamnophis_cyrtopsis_(Natricidae).jpg"><img src="../..\images/Thamnophis_cyrtopsis_(Natricidae).jpg" class="img-fluid figure-img" alt="Thamnophis cyrtopsis (Family: Natricidae). Common name: blackneck garter snake."></a></p>
<figcaption>Juan Carlos Fonseca Mata, CC BY-SA 4.0 &lt;https://creativecommons.org/licenses/by-sa/4.0&gt;, via Wikimedia Commons</figcaption>
</figure>
</div>
<p>One of their analyses considered how snake movement (labelled 0 for no movement, 1 for movement) differed among age classes (adults and juveniles).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>behavior <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/database_Thamnophis_crytopsis_study - Copy of THCY_MH_BEH.csv"</span>,</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">stringsAsFactors =</span> T)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>behavior_table <span class="ot">&lt;-</span> <span class="fu">table</span>(behavior<span class="sc">$</span>movement, behavior<span class="sc">$</span>ageclass)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>behavior_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   
    adult juvenile
  0    29       29
  1    14       16</code></pre>
</div>
</div>
<p>You may recognize this analyses could be carried out using <span class="math inline">\(\chi^2\)</span> tests.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(behavior_table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test with Yates' continuity correction

data:  behavior_table
X-squared = 0.0051228, df = 1, p-value = 0.9429</code></pre>
</div>
</div>
<p>which indicate no impact of age class on movement (p=.9429). You (hopefully) also remember we mentioned the Yate’s correction (noted in the output) was due to using a continuous distribution to fit counts. If we remove it,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(behavior_table, <span class="at">correct =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test

data:  behavior_table
X-squared = 0.087924, df = 1, p-value = 0.7668</code></pre>
</div>
</div>
<p>we find an answer that we can replicate using GLM (since the outcome is a 0 or 1 here!). We do this using the <em>glm</em> function in base R, which is very similar to the <em>lm</em> function; the only difference is we now note the distribution we are exploring using the “family” argument. Note Bernoulli data is considered a subset of binomial data and thus classified as “binomial”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>as_glm <span class="ot">&lt;-</span> <span class="fu">glm</span>(movement<span class="sc">~</span>ageclass, behavior, <span class="at">family =</span> <span class="st">"binomial"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then use our normal approach (including <em>summary</em>, <em>Anova</em>, and <em>glht</em> functions) to consider the outcomes. Checking assumptions is a little trickier than for linear models (as noted above). One major check is typically ensuring the dispersion parameter is correct. We can typically check the parameter by dividing the residual deviance by its associated degrees of freedom; this information is provided by the <em>summary</em> function (here, 112.84/86);</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(as_glm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = movement ~ ageclass, family = "binomial", data = behavior)

Coefficients:
                 Estimate Std. Error z value Pr(&gt;|z|)  
(Intercept)       -0.7282     0.3254  -2.238   0.0252 *
ageclassjuvenile   0.1335     0.4504   0.296   0.7669  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 112.93  on 87  degrees of freedom
Residual deviance: 112.84  on 86  degrees of freedom
  (15 observations deleted due to missingness)
AIC: 116.84

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
<p>For Bernoulli data like this, however, over-dispersion is not an issue <span class="citation" data-cites="molenberghs2012">(<a href="#ref-molenberghs2012" role="doc-biblioref">Molenberghs et al. 2012</a>)</span>. For binomial and Poisson data, this should be approximately one. While we can still plot the model like we did for lms</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(as_glm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-22-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-22-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-22-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>the findings may not be as useful. We only have 4 predicted values (why?) and thus only 4 residual values (why?), so we no longer expect the residuals to be normally distributed. Checking for patterns in these residuals is also complicated.</p>
<p>Another option (which also works for linear models) is offered by the <em>performance</em> package in R (another, related package not explored here, is <em>DHARMa</em>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(performance)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(as_glm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here we see no real issues with the model. We can also check for over-dispersion</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_overdispersion</span>(as_glm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># Overdispersion test

 dispersion ratio = 1.018
          p-value = 0.856</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>No overdispersion detected.</code></pre>
</div>
</div>
<p>but remember for Bernoulli data (n=1) this isn’t an issue anyway.</p>
<p>We can note the impact of given factors using <em>Anova</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(as_glm, <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III tests)

Response: movement
         LR Chisq Df Pr(&gt;Chisq)
ageclass 0.087974  1     0.7668</code></pre>
</div>
</div>
<p>which now uses likelihood-based tests by default. Similarly, the <em>summary</em> output notes details related to fitting the model (Number of Fisher scoring iterations). To interpret the coefficient from the glm, we need to consider the link function (here, the logit). If we backtransform the logit link function, we can actually say that juveniles had a <span class="math inline">\(e^\beta\)</span> greater odds of movement than adults, which is equal to</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">exp</span>(.<span class="dv">1335</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.142821</code></pre>
</div>
</div>
<p>and very close to 1 (thus the high p-value given our small sample size). Note this means a coefficient of 0 means no impact, since</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">exp</span>(<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
</div>
<p>While this example simply replicates the <span class="math inline">\(\chi^2\)</span> tests we learned earlier (deviations may arise when counts are low or high for any group, and thus <span class="math inline">\(\hat{p}\)</span> approaches extreme values and breaks the assumptions!), the power of these models comes from the ability to consider <em>multiple</em> variables (which we couldn’t do before). These models are sometimes called log-linear models.</p>
<p>To consider this approach and extend glm to binomial data, consider work by <span class="citation" data-cites="bueno2020">Bueno, Machado, and Leite (<a href="#ref-bueno2020" role="doc-biblioref">2020</a>)</span>. The authors carried out a series of experiments to determine how various factors impacted colonization of a host algae, <em>Sargassum filipendula</em>, by a small amphipod, <em>Cymadusa filosa</em>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://molecol-unesp-clp.jimdofree.com/2017/03/01/the-first-set-of-microsatellite-markers-for-the-amphipod-cymadusa-filosa/"><img src="../..\images/Cymadusa_filosa.jpg" class="img-fluid figure-img" alt="Rob Young, Centre for Biodiversity Genomics - CC - BY - NC - S"></a></p>
<figcaption>Rob Young, Centre for Biodiversity Genomics - CC - BY - NC - S</figcaption>
</figure>
</div>
<p>For one experiment, they stocked portions of aquariums with pieces of algal. Half the algae pieces had adult amphipods present(4 total; the species builds tubes to live on algae) while the other half had no adult residents. Juveniles (n=16 per patch) were initially placed on patches of natural substrate, artificial substrate, or bare areas(no substrate) in the aquarium. After 24 hours, the number of juveniles that colonized the focal algae were counted. Since the number of introduced juveniles was known, the number that did not colonize was also accounted for (raw data estimated from supplemental data given variation in sample size).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>juvenile <span class="ot">&lt;-</span> <span class="fu">read.csv</span> (<span class="st">"data/Supplemental_Data_Bueno_2020-S1-Presence of adults_lab.csv"</span>, </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">stringsAsFactors =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>since the <em>adults</em> column indicates the presence or absence of adults, let’s update it for clarity</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyr)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>juvenile<span class="sc">$</span>adults_updated <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">revalue</span>(<span class="fu">as.character</span>(juvenile<span class="sc">$</span>adults), <span class="fu">c</span>(<span class="st">"0"</span><span class="ot">=</span> <span class="st">"absent"</span>, <span class="st">"4"</span> <span class="ot">=</span><span class="st">"present"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is an example of a factorial-design with the following hypotheses:</p>
<ul>
<li>null hypothesis
<ul>
<li>H<sub>O</sub>: There is no impact on the habitat type juveniles start in on their likelihood to move</li>
<li>H<sub>O</sub>: There is no impact on adult presence in the new habitat on the likelihood of juveniles to move</li>
<li>H<sub>O</sub>: There is no interaction between the impacts of the habitat type juveniles start in and adult presence in the new habitat on the likelihood of juveniles to move</li>
</ul></li>
<li>alternative hypothesis
<ul>
<li>H<sub>A</sub>: There is an impact on the habitat type juveniles start in on their likelihood to move</li>
<li>H<sub>A</sub>: There is an impact on adult presence in the new habitat on the likelihood of juveniles to move</li>
<li>H<sub>A</sub>: There is an interaction between the impacts of the habitat type juveniles start in and adult presence in the new habitat on the likelihood of juveniles to move</li>
</ul></li>
</ul>
<p>The outcome, however, is a proportion. Note we <em>can</em> fit a linear model to the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>juvenile_fit_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(Colonized <span class="sc">~</span> adults_updated<span class="sc">*</span>source.habitat, </span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>    juvenile)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(juvenile_fit_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-30-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-30-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-30-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The assumptions don’t even look that bad, but note the residuals are over-dispersed. We could also assess the model using</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(juvenile_fit_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>which again shows some issues.</p>
<p>We also know the model will make predictions that aren’t logical (outcomes outside of the 0-1 range). For this reason we will use a generalized linear model (or a log linear model) to analyze the data. We can fit a glm using the binomial family.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>juvenile_fit_glm <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">cbind</span>(Colonized, Not_colonized) <span class="sc">~</span> adults<span class="sc">*</span>source.habitat, </span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>    juvenile, <span class="at">family =</span> <span class="st">"binomial"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The summary again allows us to estimate the dispersion parameter by dividing the residual deviance by its associated degrees of freedom, and we can now consider the impact of multiple variables (including interactions, which are significant here!).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(juvenile_fit_glm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = cbind(Colonized, Not_colonized) ~ adults * source.habitat, 
    family = "binomial", data = juvenile)

Coefficients:
                             Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)                   1.84583    0.31063   5.942 2.81e-09 ***
adults                       -0.28417    0.09507  -2.989 0.002799 ** 
source.habitatnatural        -3.88271    0.43669  -8.891  &lt; 2e-16 ***
source.habitatnone            1.02207    0.55484   1.842 0.065459 .  
adults:source.habitatnatural  0.46383    0.13819   3.356 0.000789 ***
adults:source.habitatnone    -0.38722    0.15665  -2.472 0.013441 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 347.435  on 35  degrees of freedom
Residual deviance:  97.393  on 30  degrees of freedom
AIC: 186.64

Number of Fisher Scoring iterations: 5</code></pre>
</div>
</div>
<p>The residual deviance/residual df (97.393/30) is ~ 3 when it is supposed to be 1. We’ll come back to this (and should), but just to show what we can do with these models, note we can use our <em>Anova</em> function to investigate outcomes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(juvenile_fit_glm, <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III tests)

Response: cbind(Colonized, Not_colonized)
                      LR Chisq Df Pr(&gt;Chisq)    
adults                   9.712  1   0.001831 ** 
source.habitat         195.748  2  &lt; 2.2e-16 ***
adults:source.habitat   34.738  2  2.862e-08 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>We see that there is a significant interaction between the presence of adults and source habitat on the proportion of amphipods that move to colonize algae. Given this, we can follow our typical approach to significant interactions - break the data into subsets.</p>
<p>However, we may want to reconsider this model given the over-dispersion issue. What should we do? One option is to use an approach that estimates the dispersion parameter. Though not fully developed here (but a little more below), this approach uses a “quasi-binomial” family.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>juvenile_fit_glm_quasi <span class="ot">&lt;-</span> <span class="fu">glm</span>(<span class="fu">cbind</span>(Colonized, Not_colonized) <span class="sc">~</span> adults<span class="sc">*</span>source.habitat, </span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>    juvenile, <span class="at">family =</span> <span class="st">"quasibinomial"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can investigate this model using our established protocols</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(juvenile_fit_glm_quasi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = cbind(Colonized, Not_colonized) ~ adults * source.habitat, 
    family = "quasibinomial", data = juvenile)

Coefficients:
                             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)                    1.8458     0.5562   3.319  0.00238 ** 
adults                        -0.2842     0.1702  -1.669  0.10544    
source.habitatnatural         -3.8827     0.7819  -4.966 2.56e-05 ***
source.habitatnone             1.0221     0.9934   1.029  0.31177    
adults:source.habitatnatural   0.4638     0.2474   1.875  0.07060 .  
adults:source.habitatnone     -0.3872     0.2805  -1.381  0.17760    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for quasibinomial family taken to be 3.205612)

    Null deviance: 347.435  on 35  degrees of freedom
Residual deviance:  97.393  on 30  degrees of freedom
AIC: NA

Number of Fisher Scoring iterations: 5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(juvenile_fit_glm_quasi, <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III tests)

Response: cbind(Colonized, Not_colonized)
                      LR Chisq Df Pr(&gt;Chisq)    
adults                   3.030  1   0.081751 .  
source.habitat          61.064  2  5.496e-14 ***
adults:source.habitat   10.837  2   0.004435 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Since we estimated our dispersion parameter, we can also use F tests here</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(juvenile_fit_glm_quasi, <span class="at">type=</span><span class="st">"III"</span>, <span class="at">test =</span> <span class="st">"F"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III tests)

Response: cbind(Colonized, Not_colonized)
Error estimate based on Pearson residuals 

                       Sum Sq Df F values    Pr(&gt;F)    
adults                  9.712  1   3.0298  0.091998 .  
source.habitat        195.748  2  30.5326 5.842e-08 ***
adults:source.habitat  34.738  2   5.4184  0.009796 ** 
Residuals              96.167 30                       
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</section>
<section id="poisson-regression" class="level4">
<h4 class="anchored" data-anchor-id="poisson-regression">Poisson regression</h4>
<p>Poisson regression focuses on count-based data. For simplicity (and to demonstrate the connections) here we note that we can almost always assess binomial data using the Poisson approach <em>if we only focus on the number of successes</em>. As you might imagine, this means we are ignoring data, which isn’t a great idea.</p>
<p>For example, we could consider the shrimp data like this</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>juvenile_fit_glm_poisson <span class="ot">&lt;-</span> <span class="fu">glm</span>(Colonized <span class="sc">~</span> adults<span class="sc">*</span>source.habitat, </span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>    juvenile, <span class="at">family =</span> <span class="st">"poisson"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We could use our familiar protocol to assess the model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(juvenile_fit_glm_poisson)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-39-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-39-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-39-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(juvenile_fit_glm_poisson)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-39-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_overdispersion</span>(juvenile_fit_glm_poisson)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># Overdispersion test

       dispersion ratio =  0.840
  Pearson's Chi-Squared = 25.213
                p-value =  0.715</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>No overdispersion detected.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(juvenile_fit_glm_poisson)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = Colonized ~ adults * source.habitat, family = "poisson", 
    data = juvenile)

Coefficients:
                             Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)                   2.53897    0.11471  22.134  &lt; 2e-16 ***
adults                       -0.04690    0.04260  -1.101    0.271    
source.habitatnatural        -1.84583    0.31063  -5.942 2.81e-09 ***
source.habitatnone            0.14660    0.15659   0.936    0.349    
adults:source.habitatnatural  0.16178    0.10155   1.593    0.111    
adults:source.habitatnone    -0.04885    0.05972  -0.818    0.413    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 133.282  on 35  degrees of freedom
Residual deviance:  27.223  on 30  degrees of freedom
AIC: 174.77

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
<p>Noting a lack of issues, we can try</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(juvenile_fit_glm_poisson, <span class="at">type=</span> <span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III tests)

Response: Colonized
                      LR Chisq Df Pr(&gt;Chisq)    
adults                   1.218  1     0.2698    
source.habitat          72.622  2     &lt;2e-16 ***
adults:source.habitat    4.499  2     0.1054    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Similar to a binomial model, we can exponentiate parameter values to aid interpretation, but this time it means how increasing the measured value by 1 leads to a multiplicative change in the outcome. Again, a coefficient value of 0 means no impact.</p>
<p>So which is right? Here I might argue that “losing” data in the Poisson approach justifies the binomial (or quasi-binomial) approach.</p>
<p>The Poission distribution also assumes a dispersion parameter of 1. When this is not true, options include using a quasi- approach (quasipossion) or a negative binomial GLM.</p>
</section>
<section id="negative-binomial-regression" class="level4">
<h4 class="anchored" data-anchor-id="negative-binomial-regression">Negative binomial regression</h4>
<p>Similar to poisson, but an additional parameter allows the mean to not be equal to the variance. Packages to build these models include <em>MASS</em>(glm.nb function).</p>
</section>
<section id="gamma-regression" class="level4">
<h4 class="anchored" data-anchor-id="gamma-regression">Gamma regression</h4>
<p>For an example of continuous data using glms, consider the work by <span class="citation" data-cites="li2024">Li et al. (<a href="#ref-li2024" role="doc-biblioref">2024</a>)</span> on how humidity impacted responses to heat stress in parasitoids (<em>Venturia canescens</em>) and their hosts (<em>Plodia interpunctella</em>, Indian meal moths). Eggs were placed in incubators kept at ambient or increased humidity levels. Half of each group was parasitized during their 4th instar. Parasitized eggs were then exposed to no heat stress or stress for 6 or 72 hours in their 4th or 5th instar phase. Multiple traits in host and parasitoid were then measured.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../..\images/Venturia_canescens_lateral2.jpg" class="img-fluid figure-img" alt="Venturia canescens"></p>
<figcaption>Parasitic wasp ; image by Slimguy, CC BY 4.0 &lt;https://creativecommons.org/licenses/by/4.0&gt;, via Wikimedia Commons</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../..\images/Plodia_interpunctella_lrv_79.jpg" class="img-fluid figure-img" alt="Plodia interpunctella"></p>
<figcaption>Indian meal moth larvae ; Public domain via Wikimedia Commons</figcaption>
</figure>
</div>
<p>One measured factor was wasp development time.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>wasp_no_humidity <span class="ot">&lt;-</span> <span class="fu">read.csv</span> (<span class="st">"data/Nonhumidity_hostwithparasitoids.csv"</span>, </span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">stringsAsFactors =</span> T)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>wasp_humidity <span class="ot">&lt;-</span> <span class="fu">read.csv</span> (<span class="st">"data/humidity_hostwithparasitoids.csv"</span>, </span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">stringsAsFactors =</span> T)</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>wasp <span class="ot">&lt;-</span> <span class="fu">rbind</span>(wasp_humidity,wasp_no_humidity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in `[&lt;-.factor`(`*tmp*`, ri, value = c(NA, NA, 20230426L, 20230425L, :
invalid factor level, NA generated</code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>wasp<span class="sc">$</span>Parasited_date <span class="ot">&lt;-</span><span class="fu">as.Date</span>(wasp<span class="sc">$</span>Parasited_date, <span class="st">"%Y%m%d"</span>) </span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>wasp<span class="sc">$</span>Emerge_date <span class="ot">&lt;-</span><span class="fu">as.Date</span>(<span class="fu">as.character</span>(wasp<span class="sc">$</span>Emerge_date), <span class="st">"%Y%m%d"</span>) </span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>wasp<span class="sc">$</span>develop_time <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(wasp<span class="sc">$</span>Emerge_date <span class="sc">-</span> wasp<span class="sc">$</span>Parasited_date)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The experimental data can be visualized as:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyr)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Hmisc)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>wasp<span class="sc">$</span>Duration_new <span class="ot">&lt;-</span> <span class="fu">revalue</span>(wasp<span class="sc">$</span>Duration, <span class="fu">c</span>(<span class="st">"C"</span><span class="ot">=</span><span class="st">"0"</span>, <span class="st">"L"</span><span class="ot">=</span><span class="st">"72"</span>, <span class="st">"S"</span><span class="ot">=</span><span class="st">"6"</span>))</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>wasp<span class="sc">$</span>Duration_new <span class="ot">&lt;-</span> <span class="fu">factor</span>(wasp<span class="sc">$</span>Duration_new, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"0"</span>, <span class="st">"6"</span>, <span class="st">"72"</span>))</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>wasp<span class="sc">$</span>Humidity_new <span class="ot">&lt;-</span> <span class="fu">revalue</span>(wasp<span class="sc">$</span>Humidity, <span class="fu">c</span>(<span class="st">"H"</span><span class="ot">=</span><span class="st">"high"</span>, <span class="st">"NH"</span><span class="ot">=</span><span class="st">"ambient"</span>))</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>wasp<span class="sc">$</span>Humidity_new <span class="ot">&lt;-</span> <span class="fu">factor</span>(wasp<span class="sc">$</span>Humidity_new, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"ambient"</span>, <span class="st">"high"</span>))</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(wasp, <span class="fu">aes</span>(<span class="at">x=</span>Duration_new, <span class="at">y=</span>develop_time, <span class="at">color=</span>Humidity_new))<span class="sc">+</span></span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>()<span class="sc">+</span></span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Heat Stress Duration (hours)"</span>, <span class="at">y=</span> <span class="st">"Development time (days)"</span>, </span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">color=</span><span class="st">"Humidity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 404 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This could be analyzed using a linear model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>time_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(develop_time<span class="sc">~</span>Humidity<span class="sc">*</span>Duration, wasp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>but the residuals are clearly overdispersed and may be increasing with the mean.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(time_model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-44-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-44-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-44-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This makes sense as the data is bounded. Negative development times are not possible, and we might assume that as the average development date increases, there is more room for variability (so variance is not constant). A gamma distribution can help these issues.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>time_model_gamma <span class="ot">&lt;-</span> <span class="fu">glm</span>(develop_time<span class="sc">~</span>Humidity<span class="sc">*</span>Duration, wasp, <span class="at">family =</span> Gamma)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(time_model_gamma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>and AIC suggests it is very slightly better (remember, linear models are very robust!).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">AICc</span>(time_model, time_model_gamma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 df     AICc
time_model        7 970.7620
time_model_gamma  7 970.0118</code></pre>
</div>
</div>
<p>We can then analyze the impact of various factors on development time using our standard approach.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(time_model_gamma, <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III tests)

Response: develop_time
                  LR Chisq Df Pr(&gt;Chisq)    
Humidity            2.3234  1     0.1274    
Duration           18.9323  2  7.743e-05 ***
Humidity:Duration   3.5064  2     0.1732    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Since there is not evidence for a significant interaction, we could remove it</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>time_model_gamma <span class="ot">&lt;-</span> <span class="fu">glm</span>(develop_time<span class="sc">~</span>Humidity<span class="sc">+</span>Duration, wasp, <span class="at">family =</span> Gamma)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(time_model_gamma, <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III tests)

Response: develop_time
         LR Chisq Df Pr(&gt;Chisq)    
Humidity   0.0908  1     0.7632    
Duration  24.1852  2  5.601e-06 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Results indicate heat stress, but not humidity, impact wasp development time. Since there are three levels, we could use post hoc tests to determine which treatments differ from others.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">glht</span>(time_model_gamma, <span class="at">linfct =</span> <span class="fu">mcp</span>(<span class="st">"Duration"</span><span class="ot">=</span> <span class="st">"Tukey"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Simultaneous Tests for General Linear Hypotheses

Multiple Comparisons of Means: Tukey Contrasts


Fit: glm(formula = develop_time ~ Humidity + Duration, family = Gamma, 
    data = wasp)

Linear Hypotheses:
            Estimate Std. Error z value Pr(&gt;|z|)    
L - C == 0 -0.007440   0.001938  -3.839 0.000357 ***
S - C == 0 -0.001461   0.001912  -0.764 0.721799    
S - L == 0  0.005979   0.001373   4.355  &lt; 1e-04 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
(Adjusted p values reported -- single-step method)</code></pre>
</div>
</div>
<p>This suggests the longest duration is signifcantly different than the others and has a lower value. Similary, let’s consider the coefficients.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(time_model_gamma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = develop_time ~ Humidity + Duration, family = Gamma, 
    data = wasp)

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  0.0505499  0.0018403  27.468  &lt; 2e-16 ***
HumidityNH   0.0003849  0.0012771   0.301 0.763501    
DurationL   -0.0074400  0.0019381  -3.839 0.000174 ***
DurationS   -0.0014607  0.0019124  -0.764 0.446030    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for Gamma family taken to be 0.03109605)

    Null deviance: 6.2813  on 175  degrees of freedom
Residual deviance: 5.5285  on 172  degrees of freedom
  (404 observations deleted due to missingness)
AIC: 968.81

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
<p>Coefficients indicate the highest predicted value for control conditions (remember, that’s the intercept condition), followed short and long stress events. This would suggest that heat stress <em>lowers</em> development time…but remember the graphed data (here with humidity impacts removed).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(wasp, <span class="fu">aes</span>(<span class="at">x=</span>Duration_new, <span class="at">y=</span>develop_time))<span class="sc">+</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>()<span class="sc">+</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun.data=</span><span class="st">"mean_sdl"</span>, <span class="at">geom=</span><span class="st">"crossbar"</span>, <span class="at">width=</span><span class="fl">0.5</span>)<span class="sc">+</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">"Heat Stress Duration (hours)"</span>, <span class="at">y=</span> <span class="st">"Development time (days)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 404 rows containing non-finite outside the scale range
(`stat_summary()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 404 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here we see that the highest heat stress level is associated with a longer wasp development time. Why do the results seem contradictory?</p>
<p>Again, this is due to the link function. For gamma models, the link is the inverse, so to determine the impact of a given coefficient on the raw data we would combine the intercept and coefficients in this way.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">/</span>(.<span class="dv">05</span><span class="fl">-.007</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 23.25581</code></pre>
</div>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">/</span>(.<span class="dv">05</span><span class="fl">-.001</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 20.40816</code></pre>
</div>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">/</span>(.<span class="dv">05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 20</code></pre>
</div>
</div>
<p>Which indicates a longer development time (23 days vs 20) for wasps exposed to more heat stress.</p>
</section>
<section id="beta-regression" class="level4">
<h4 class="anchored" data-anchor-id="beta-regression">Beta regression</h4>
<p>Beta regression <span class="citation" data-cites="geissinger2022">(<a href="#ref-geissinger2022" role="doc-biblioref">Geissinger et al. 2022</a>)</span> is similar to logistic/binomial regression, but it focuses on true proportion data. In other words, whereas binomial regression focuses on modeling the number of successes in a given number of trials, beta regression focuses on the distribution of the probability parameter itself. If we are exploring this, it means we know the value can’t really be 0 or 1 (or we would have no variation), so responses may not be 0 or 1 (they must be in this range). Since “accidents” do happen and may lead to zeroes or ones in the data, there are options for adjusting the data to account for these issues <span class="citation" data-cites="corlatti2021">(p. 11, <a href="#ref-corlatti2021" role="doc-biblioref">Corlatti 2021</a>)</span>. This is useful if the outcome is between 0 and 1 (e.g.&nbsp;% cover, % nitrogen). Packages that can do beta analysis in R include <em>betareg</em> and <em>glmmTMB</em>.</p>
</section>
<section id="non-linear-options" class="level4">
<h4 class="anchored" data-anchor-id="non-linear-options">Non-linear options</h4>
<p>All the options above still rely on a linear-link property: the systematic component of the model has a linear relationship to a predictor (in glm, we just transformed this back to the raw data as needed). However, we may instead actually want a non-linear relationship between our predictors and outcome. In fact, we may actually want to compare various linear and non-linear fits <span class="citation" data-cites="dornberger2016">(e.g., <a href="#ref-dornberger2016" role="doc-biblioref">Dornberger et al. 2016</a>)</span>or determine parameter values for a specified relationship.</p>
<p>For example, many fisheries use length of organisms as baseline data because it’s easy to collect, but we may want to use length data to estimate mass. Determining this length-weight relationship may be useful to fisheries management.</p>
<p>We know from basic principles the relationship between length and mass is probably not linear, because length is only capturing growth in one dimension, whereas mass is likely more closely related to volume. Given this, we might decide to fit the equation</p>
<p><span class="math display">\[
Weight =a*Length^B
\]</span> to determine this relationship.</p>
<p>A group <span class="citation" data-cites="gosnell2023">(<a href="#ref-gosnell2023" role="doc-biblioref">Gosnell et al. 2023</a>)</span> used data collected on whelks from multiple locations to consider the parameter values for this model (only 2 locations used here).</p>
<p><a href="https://commons.wikimedia.org/wiki/File:Kellet%27s_Whelk_-_Kelletia_kelletii_(29525885698).jpg"><img src="../..\images/1024px-Kellet's_Whelk_-_Kelletia_kelletii_(29525885698).jpg" class="img-fluid" alt="Björn S..., CC BY-SA 2.0 <https://creativecommons.org/licenses/by-sa/2.0>, via Wikimedia Commons"></a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>whelk <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"https://raw.githubusercontent.com/jsgosnell/CUNY-BioStats/master/datasets/whelk.csv"</span>) </span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(whelk, <span class="fu">aes</span>(<span class="at">x=</span>Shell.Length, <span class="at">y =</span> Mass, <span class="at">color=</span>Location)) <span class="sc">+</span> </span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span> <span class="st">"Mass (g)"</span>, <span class="at">x=</span> <span class="st">"Shell length (mm)"</span>, <span class="at">color=</span> <span class="st">"Population"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 61 rows containing missing values or values outside the scale range
(`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>whelk_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(Mass <span class="sc">~</span> Shell.Length, whelk, <span class="at">na.action =</span> na.omit)</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>whelk_nls <span class="ot">&lt;-</span> <span class="fu">nls</span>(Mass <span class="sc">~</span> b0 <span class="sc">*</span> Shell.Length<span class="sc">^</span>b1, whelk, </span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">start =</span> <span class="fu">list</span>(<span class="at">b0 =</span> <span class="dv">1</span>, <span class="at">b1=</span><span class="dv">3</span>), <span class="at">na.action =</span> na.omit)</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a><span class="fu">AICc</span>(whelk_lm, whelk_nls)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          df     AICc
whelk_lm   3 3947.579
whelk_nls  3 3800.106</code></pre>
</div>
</div>
<p>The non-linear fit is obviously better. Parameters can be found using</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(whelk_nls)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Formula: Mass ~ b0 * Shell.Length^b1

Parameters:
    Estimate Std. Error t value Pr(&gt;|t|)    
b0 2.134e-04  5.737e-05   3.719 0.000228 ***
b1 2.855e+00  5.576e-02  51.195  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 24.24 on 410 degrees of freedom

Number of iterations to convergence: 5 
Achieved convergence tolerance: 7.614e-07
  (61 observations deleted due to missingness)</code></pre>
</div>
</div>
<p>Comparing parameters among sites is slightly more difficult, as dummy variables need to be included for each site; alternatlvely, data can be transformed and a linear model considered.</p>
<p>Notice this approach is mechanistic - we set the relationship and found the best parameters. There are other ways to build non-linear relationships, however, and some allow algorithms to find the best relationship. While these may not be hypothesis-driven, they may prove very useful for predictions. One example of this approach is known as generalized additive modelings (or gams). These models can fit multiple functions to the data, including piece-wise functions. In this example we consider how a splines, or smooth, polynomial curves that are fit to various portions of the data and then “knotted” together, fit the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>whelk_gam_spline <span class="ot">&lt;-</span> <span class="fu">gam</span>(Mass <span class="sc">~</span> <span class="fu">s</span>(Shell.Length), <span class="at">data =</span> whelk)</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(whelk_gam_spline)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Family: gaussian 
Link function: identity 

Formula:
Mass ~ s(Shell.Length)

Parametric coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  147.321      1.198     123   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Approximate significance of smooth terms:
                 edf Ref.df     F p-value    
s(Shell.Length) 3.98  4.993 807.4  &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

R-sq.(adj) =  0.908   Deviance explained = 90.8%
GCV = 598.47  Scale est. = 591.24    n = 412</code></pre>
</div>
</div>
<p>These models have the same assumptions as linear models <span class="citation" data-cites="zuur2009">(p.&nbsp;58, <a href="#ref-zuur2009" role="doc-biblioref">A. Zuur et al. 2009</a>)</span>.</p>
<p>As noted above, this is just meant to introduce these approaches. We can compare these (and others, such as loess) using</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">AICc</span>(whelk_gam_spline, whelk_nls, whelk_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                       df     AICc
whelk_gam_spline 5.979836 3805.835
whelk_nls        3.000000 3800.106
whelk_lm         3.000000 3947.579</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="broken-assumption-data-are-not-independent" class="level2">
<h2 class="anchored" data-anchor-id="broken-assumption-data-are-not-independent">Broken assumption: Data are not independent</h2>
<section id="in-respect-to-predictors" class="level3">
<h3 class="anchored" data-anchor-id="in-respect-to-predictors">In respect to predictors</h3>
<p>A major issue for linear models is when predictors are co-linear. Mathematically speaking, perfect collinearity occurs when any column of the design (<em>X</em>) matrix can be derived by combining other columns. Perfect collinearity will lead to a message noting singularity issues, which is R’s way of telling you the matrix isn’t invertible (which it has to be to solve the equation).</p>
<p>Even partial collinearity will lead to an increase in Type II errors <span class="citation" data-cites="zuur2010">(<a href="#ref-zuur2010" role="doc-biblioref">A. F. Zuur, Ieno, and Elphick 2010</a>)</span>. To put it simply, partitioning variance among related predictors is hard. For this reason, a few approaches may be used.</p>
<section id="check-for-issues" class="level4">
<h4 class="anchored" data-anchor-id="check-for-issues">Check for issues</h4>
<p>The first step is identifying issues. From the outset, relationships among predictor variables can be assessed using the <em>pairs</em> function in R. If two variables are highly correlated (r<sup>2</sup> &gt; .8 is a general limit), only one should be included in the model. Similarly, variance inflation factors (vif) can be assessed for the final and other models to consider this issues (all this is covered in the <a href="../../content/chapters/Combining_numerical_and_categorical_predictors.html">previous chapter that introduces multiple regression</a>. If this occurs, some variables may need to be removed from consideration.</p>
</section>
<section id="multivariate-methods" class="level4">
<h4 class="anchored" data-anchor-id="multivariate-methods">Multivariate methods</h4>
<p>Covered in later chapters, but if predictors (or multiple outcomes) are correlated, methods can be used to derive statistically independent predictors (though they may be hard to interpret).</p>
</section>
</section>
<section id="in-respect-to-measured-outcomes" class="level3">
<h3 class="anchored" data-anchor-id="in-respect-to-measured-outcomes">In respect to measured outcomes</h3>
<p>When outcome variables are linked together, a few options exist. Note this issue may be obvious from checks of assumptions, but it also may be due to experimental design.</p>
<section id="include-random-effects" class="level4">
<h4 class="anchored" data-anchor-id="include-random-effects">Include random effects</h4>
<p>We start with an approach that models connections among variables via random effects.</p>
<p>Consider this toy example. In order to investigate impacts of predator stress on oysters, oysters are exposed to predator cues by dropping in water from tanks with 0, low (.25/m2), or high (2/m2) densities of predators. After two months, changes in oyster shell length (growth) was measured. Here we make the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>experiment <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(                      <span class="at">predator_cue =</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"none"</span>,<span class="dv">25</span>),</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>                                               <span class="fu">rep</span>(<span class="st">"low"</span>,<span class="dv">25</span>),</span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>                                               <span class="fu">rep</span>(<span class="st">"high"</span>,<span class="dv">25</span>)),<span class="dv">1</span>))),</span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>                         <span class="at">growth =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">25</span>,<span class="dv">5</span>,.<span class="dv">15</span>),</span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">rnorm</span>(<span class="dv">25</span>, <span class="fl">4.7</span>, .<span class="dv">15</span>),</span>
<span id="cb111-7"><a href="#cb111-7" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">rnorm</span>(<span class="dv">25</span>, <span class="fl">4.4</span>, .<span class="dv">15</span>)))</span>
<span id="cb111-8"><a href="#cb111-8" aria-hidden="true" tabindex="-1"></a>experiment[experiment<span class="sc">$</span>growth<span class="sc">&lt;</span><span class="dv">0</span>, <span class="st">"growth"</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The data look something like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>experiment<span class="sc">$</span>predator_cue <span class="ot">&lt;-</span> <span class="fu">factor</span>(experiment<span class="sc">$</span>predator_cue, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"none"</span>, <span class="st">"low"</span>, <span class="st">"high"</span>))</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(experiment, <span class="fu">aes</span>(<span class="at">x=</span>predator_cue, <span class="at">y=</span>growth, <span class="at">color=</span> predator_cue))<span class="sc">+</span></span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span> <span class="st">"Predator exposure"</span>, <span class="at">y=</span> <span class="st">"Growth (mm)"</span>)<span class="sc">+</span></span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>   <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-59-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>How would you analyze this data? If we assume that each oyster was in its own space (and received cues from a unique predator) and is thus an independent measurement, you hopefully recognize this as a lm/ANOVA that you know how to analyze. If you need a reminder, see the <a href="../../content/chapters/Compare_means_among_populations.html">chapter on ANOVAs.</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>growth_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(growth<span class="sc">~</span>predator_cue, experiment)</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(growth_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-60-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-60-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-60-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-60-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(growth_lm, <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: growth
             Sum Sq Df  F value    Pr(&gt;F)    
(Intercept)  627.97  1 26982.57 &lt; 2.2e-16 ***
predator_cue   4.84  2   103.99 &lt; 2.2e-16 ***
Residuals      1.68 72                       
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>This experimental approach takes a lot of space (and time!). If I have a large tank, I may instead decide to put a few more oysters in there. At an extreme, let’s instead use one tank was used for each predator level, with twenty-five oysters measured in each tank.However, this means our measurements are connected; remember <a href="../../content/chapters/More_ANOVAs.html">blocking and paired tests</a>? So our data may look more like</p>
<!-- helpful here https://verynormal.substack.com/p/office-hours-1-simulating-from-mixed; https://aosmith.rbind.io/2018/04/23/simulate-simulate-part-2/ -->
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>n_levels <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>replicates_each_level <span class="ot">&lt;-</span> <span class="dv">1</span> </span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>J <span class="ot">=</span> n_levels <span class="sc">*</span> replicates_each_level             <span class="co"># number of tanks</span></span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>n_j <span class="ot">=</span> <span class="dv">25</span>              <span class="co"># number of observations per tank </span></span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>sigma_b <span class="ot">=</span> .<span class="dv">15</span>           <span class="co"># variance of the random intercept</span></span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>sigma_e <span class="ot">=</span> .<span class="dv">15</span>           <span class="co"># variance of the random noise</span></span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>fixed_effect <span class="ot">=</span> <span class="sc">-</span>.<span class="dv">05</span>      <span class="co"># population-level effect</span></span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>fixed_intercept <span class="ot">=</span> <span class="dv">5</span>   <span class="co"># population-level intercept</span></span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>experiment_with_3_tanks <span class="ot">&lt;-</span>  <span class="fu">data.frame</span>(</span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">tank =</span> <span class="fu">rep</span>(<span class="fu">sample</span>(letters[<span class="dv">1</span><span class="sc">:</span>J]),<span class="at">each=</span>n_j), <span class="co"># puts these in random order!</span></span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">b_i =</span> <span class="fu">rep</span>(<span class="fu">rnorm</span>(J, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> sigma_b), <span class="at">each=</span>n_j),</span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Generate predictor </span></span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">X =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">each=</span> n_j<span class="sc">*</span>J<span class="sc">/</span><span class="dv">3</span>))</span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-17"><a href="#cb116-17" aria-hidden="true" tabindex="-1"></a>experiment_with_3_tanks<span class="sc">$</span>predator_cue <span class="ot">&lt;-</span> <span class="fu">revalue</span>(<span class="fu">as.character</span>(experiment_with_3_tanks<span class="sc">$</span>X), <span class="fu">c</span>(<span class="st">"0"</span> <span class="ot">=</span> <span class="st">"none"</span>, </span>
<span id="cb116-18"><a href="#cb116-18" aria-hidden="true" tabindex="-1"></a>                                                                             <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"low"</span>, </span>
<span id="cb116-19"><a href="#cb116-19" aria-hidden="true" tabindex="-1"></a>                                                                             <span class="st">"2"</span><span class="ot">=</span><span class="st">"high"</span>))</span>
<span id="cb116-20"><a href="#cb116-20" aria-hidden="true" tabindex="-1"></a>experiment_with_3_tanks<span class="sc">$</span>predator_cue <span class="ot">&lt;-</span> <span class="fu">factor</span>(experiment_with_3_tanks<span class="sc">$</span>predator_cue, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"none"</span>, <span class="st">"low"</span>, <span class="st">"high"</span>))</span>
<span id="cb116-21"><a href="#cb116-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-22"><a href="#cb116-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate outcome from predictor &amp; subject-specific intercept</span></span>
<span id="cb116-23"><a href="#cb116-23" aria-hidden="true" tabindex="-1"></a>experiment_with_3_tanks<span class="sc">$</span>growth <span class="ot">&lt;-</span> fixed_intercept <span class="sc">+</span> experiment_with_3_tanks<span class="sc">$</span>b_i <span class="sc">+</span> fixed_effect<span class="sc">*</span> experiment_with_3_tanks<span class="sc">$</span>X <span class="sc">+</span> <span class="fu">rnorm</span>(n_j<span class="sc">*</span>J, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> sigma_e)</span>
<span id="cb116-24"><a href="#cb116-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-25"><a href="#cb116-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-26"><a href="#cb116-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-27"><a href="#cb116-27" aria-hidden="true" tabindex="-1"></a>experiment_with_3_tanks <span class="ot">&lt;-</span> experiment_with_3_tanks[<span class="fu">order</span>(experiment_with_3_tanks<span class="sc">$</span>tank),]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now the data look like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(experiment_with_3_tanks, <span class="fu">aes</span>(<span class="at">x=</span>predator_cue, <span class="at">y=</span>growth, <span class="at">color=</span> predator_cue))<span class="sc">+</span></span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span> <span class="st">"Predator exposure"</span>, <span class="at">y=</span> <span class="st">"Growth (mm)"</span>)<span class="sc">+</span></span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a>   <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now we are stuck with a question. Do I have n=75 (the number of oysters) or n=3 (the number of tanks) for the experiment? Ignoring linkages leads to pseudopreplication, an issue that has been noted in ecology and other fields <span class="citation" data-cites="hurlbert1984 heffner1996 lazic2010">(<a href="#ref-hurlbert1984" role="doc-biblioref">Hurlbert 1984</a>; <a href="#ref-heffner1996" role="doc-biblioref">Heffner, Butler, and Reilly 1996</a>; <a href="#ref-lazic2010" role="doc-biblioref">Lazic 2010</a>)</span>.</p>
<p>There are several ways to deal with this. Here we explore each for our oyster example.</p>
<section id="find-average-for-each-unit" class="level5">
<h5 class="anchored" data-anchor-id="find-average-for-each-unit">Find average for each unit</h5>
<p>One way of doing this focuses on the average for each unit</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Rmisc)</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>averaged_experiment_3_tanks <span class="ot">&lt;-</span> <span class="fu">summarySE</span>(experiment_with_3_tanks, <span class="at">measurevar =</span> <span class="st">"growth"</span>,</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">groupvars =</span> <span class="fu">c</span>(<span class="st">"predator_cue"</span>, <span class="st">"tank"</span>))</span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rmarkdown)</span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a><span class="fu">paged_table</span>(averaged_experiment_3_tanks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["predator_cue"],"name":[1],"type":["fct"],"align":["left"]},{"label":["tank"],"name":[2],"type":["chr"],"align":["left"]},{"label":["N"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["growth"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["sd"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["se"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["ci"],"name":[7],"type":["dbl"],"align":["right"]}],"data":[{"1":"none","2":"a","3":"25","4":"4.819073","5":"0.1201171","6":"0.02402343","7":"0.04958192"},{"1":"low","2":"b","3":"25","4":"4.868584","5":"0.1440173","6":"0.02880347","7":"0.05944744"},{"1":"high","2":"c","3":"25","4":"4.935727","5":"0.1346676","6":"0.02693353","7":"0.05558806"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>and use that for your analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>average_analysis_3_tanks <span class="ot">&lt;-</span> <span class="fu">lm</span>(growth<span class="sc">~</span>predator_cue, averaged_experiment_3_tanks)</span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(average_analysis_3_tanks, <span class="at">type =</span> <span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in Anova.lm(average_analysis_3_tanks, type = "III"): residual df = 0</code></pre>
</div>
</div>
<p>but that leads to an issue! Since we only get 3 average outcomes and our model requires 3 degrees of freedom (consider why), we are left with no “noise” to make the denominator for our F ratio! This is partially due to a lack of replication - we’ll fix that now.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">19</span>)</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>n_levels <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>replicates_each_level <span class="ot">&lt;-</span> <span class="dv">4</span> </span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>J <span class="ot">=</span> n_levels <span class="sc">*</span> replicates_each_level             <span class="co"># number of tanks</span></span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>n_j <span class="ot">=</span> <span class="dv">25</span>              <span class="co"># number of observations per tank </span></span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>sigma_b <span class="ot">=</span> .<span class="dv">15</span>           <span class="co"># variance of the random intercept</span></span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a>sigma_e <span class="ot">=</span> .<span class="dv">15</span>           <span class="co"># variance of the random noise</span></span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a>fixed_effect <span class="ot">=</span> <span class="sc">-</span>.<span class="dv">05</span>      <span class="co"># population-level effect</span></span>
<span id="cb121-9"><a href="#cb121-9" aria-hidden="true" tabindex="-1"></a>fixed_intercept <span class="ot">=</span> <span class="dv">5</span>   <span class="co"># population-level intercept</span></span>
<span id="cb121-10"><a href="#cb121-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-11"><a href="#cb121-11" aria-hidden="true" tabindex="-1"></a>experiment_with_12_tanks <span class="ot">&lt;-</span>  <span class="fu">data.frame</span>(</span>
<span id="cb121-12"><a href="#cb121-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">tank =</span> <span class="fu">rep</span>(<span class="fu">sample</span>(letters[<span class="dv">1</span><span class="sc">:</span>J]),<span class="at">each=</span>n_j), <span class="co"># puts these in random order!</span></span>
<span id="cb121-13"><a href="#cb121-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">b_i =</span> <span class="fu">rep</span>(<span class="fu">rnorm</span>(J, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> sigma_b), <span class="at">each=</span>n_j),</span>
<span id="cb121-14"><a href="#cb121-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Generate predictor </span></span>
<span id="cb121-15"><a href="#cb121-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">X =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">each=</span> n_j<span class="sc">*</span>J<span class="sc">/</span><span class="dv">3</span>))</span>
<span id="cb121-16"><a href="#cb121-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-17"><a href="#cb121-17" aria-hidden="true" tabindex="-1"></a>experiment_with_12_tanks<span class="sc">$</span>predator_cue <span class="ot">&lt;-</span> <span class="fu">revalue</span>(<span class="fu">as.character</span>(experiment_with_12_tanks<span class="sc">$</span>X), <span class="fu">c</span>(<span class="st">"0"</span> <span class="ot">=</span> <span class="st">"none"</span>, </span>
<span id="cb121-18"><a href="#cb121-18" aria-hidden="true" tabindex="-1"></a>                                                                             <span class="st">"1"</span> <span class="ot">=</span> <span class="st">"low"</span>, </span>
<span id="cb121-19"><a href="#cb121-19" aria-hidden="true" tabindex="-1"></a>                                                                             <span class="st">"2"</span><span class="ot">=</span><span class="st">"high"</span>))</span>
<span id="cb121-20"><a href="#cb121-20" aria-hidden="true" tabindex="-1"></a>experiment_with_12_tanks<span class="sc">$</span>predator_cue <span class="ot">&lt;-</span> <span class="fu">factor</span>(experiment_with_12_tanks<span class="sc">$</span>predator_cue, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"none"</span>, <span class="st">"low"</span>, <span class="st">"high"</span>))</span>
<span id="cb121-21"><a href="#cb121-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-22"><a href="#cb121-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate outcome from predictor &amp; subject-specific intercept</span></span>
<span id="cb121-23"><a href="#cb121-23" aria-hidden="true" tabindex="-1"></a>experiment_with_12_tanks<span class="sc">$</span>growth <span class="ot">&lt;-</span> fixed_intercept <span class="sc">+</span> experiment_with_12_tanks<span class="sc">$</span>b_i <span class="sc">+</span> fixed_effect<span class="sc">*</span> experiment_with_12_tanks<span class="sc">$</span>X <span class="sc">+</span> <span class="fu">rnorm</span>(n_j<span class="sc">*</span>J, <span class="at">mean =</span> <span class="dv">0</span>, <span class="at">sd =</span> sigma_e)</span>
<span id="cb121-24"><a href="#cb121-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-25"><a href="#cb121-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-26"><a href="#cb121-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-27"><a href="#cb121-27" aria-hidden="true" tabindex="-1"></a>experiment_with_12_tanks <span class="ot">&lt;-</span> experiment_with_12_tanks[<span class="fu">order</span>(experiment_with_12_tanks<span class="sc">$</span>tank),]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can find the average for each of our 12 tanks.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>averaged_experiment_12_tanks <span class="ot">&lt;-</span> <span class="fu">summarySE</span>(experiment_with_12_tanks, <span class="at">measurevar =</span> <span class="st">"growth"</span>,</span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">groupvars =</span> <span class="fu">c</span>(<span class="st">"predator_cue"</span>, <span class="st">"tank"</span>))</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a><span class="fu">paged_table</span>(averaged_experiment_12_tanks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["predator_cue"],"name":[1],"type":["fct"],"align":["left"]},{"label":["tank"],"name":[2],"type":["chr"],"align":["left"]},{"label":["N"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["growth"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["sd"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["se"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["ci"],"name":[7],"type":["dbl"],"align":["right"]}],"data":[{"1":"none","2":"c","3":"25","4":"5.046751","5":"0.1306709","6":"0.02613419","7":"0.05393831"},{"1":"none","2":"e","3":"25","4":"5.133726","5":"0.1562976","6":"0.03125953","7":"0.06451650"},{"1":"none","2":"f","3":"25","4":"5.213888","5":"0.1376685","6":"0.02753371","7":"0.05682678"},{"1":"none","2":"j","3":"25","4":"5.241514","5":"0.1404080","6":"0.02808160","7":"0.05795757"},{"1":"low","2":"a","3":"25","4":"5.091512","5":"0.1847065","6":"0.03694129","7":"0.07624309"},{"1":"low","2":"b","3":"25","4":"5.196151","5":"0.1622862","6":"0.03245724","7":"0.06698845"},{"1":"low","2":"i","3":"25","4":"4.986351","5":"0.1039809","6":"0.02079618","7":"0.04292121"},{"1":"low","2":"l","3":"25","4":"4.961006","5":"0.1320602","6":"0.02641204","7":"0.05451178"},{"1":"high","2":"d","3":"25","4":"4.930482","5":"0.1211679","6":"0.02423358","7":"0.05001564"},{"1":"high","2":"g","3":"25","4":"5.077042","5":"0.1413902","6":"0.02827805","7":"0.05836302"},{"1":"high","2":"h","3":"25","4":"5.010585","5":"0.1517568","6":"0.03035137","7":"0.06264214"},{"1":"high","2":"k","3":"25","4":"4.937518","5":"0.1122247","6":"0.02244495","7":"0.04632409"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>plot …</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(averaged_experiment_12_tanks, <span class="fu">aes</span>(<span class="at">x=</span>predator_cue, <span class="at">y=</span>growth, <span class="at">color=</span> predator_cue))<span class="sc">+</span></span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span> <span class="st">"Predator exposure"</span>, <span class="at">y=</span> <span class="st">"Growth (mm)"</span>)<span class="sc">+</span></span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a>   <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-67-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>and analyze them.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>average_analysis_12_tanks <span class="ot">&lt;-</span> <span class="fu">lm</span>(growth<span class="sc">~</span>predator_cue, averaged_experiment_12_tanks)</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(average_analysis_12_tanks, <span class="at">type =</span> <span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: growth
              Sum Sq Df    F value    Pr(&gt;F)    
(Intercept)  106.460  1 13289.0271 1.412e-15 ***
predator_cue   0.058  2     3.6485   0.06912 .  
Residuals      0.072  9                         
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>This is one way to manage dependent data, but you have reduced your data to a much smaller number of points and are not getting credit for all your work!</p>
</section>
<section id="blocking" class="level5">
<h5 class="anchored" data-anchor-id="blocking">Blocking</h5>
<p>The blocking approach we’ve already covered works might seem appropriate here.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>blocking_analysis <span class="ot">&lt;-</span> <span class="fu">lm</span>(growth<span class="sc">~</span>predator_cue<span class="sc">+</span>tank, experiment_with_12_tanks)</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(blocking_analysis, <span class="at">type =</span> <span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in Anova_III_lm(mod, error, singular.ok = singular.ok, ...): there are aliased coefficients in the model</code></pre>
</div>
</div>
<p>but it doesn’t work. Why? This error means our model matrix has collinearity issues. We can actually see where</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="fu">alias</span>(blocking_analysis)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Model :
growth ~ predator_cue + tank

Complete :
      (Intercept) predator_cuelow predator_cuehigh tankb tankc tankd tanke
tankj  1          -1              -1                0    -1     0    -1   
tankk  0           0               1                0     0    -1     0   
      tankf tankg tankh tanki tankl
tankj -1     0     0     0     0   
tankk  0    -1    -1     0     0   </code></pre>
</div>
</div>
<p>though the output is confusing. In general, the issue here is each container only measures one level of predator risk, so risk and container are completely confounded. It’s because the container variable can fully “explain” the predation levels - this goes back to the singularity issues we discussed in introducing linear models. This is because each container only contributes to one level of other traits. If instead each unit contributed to multiple levels, like in our <a href="../../content/chapters/More_ANOVAs.html">feather experiment</a>, as show below) this isn’t an issue.</p>
</section>
<section id="random-effects" class="level5">
<h5 class="anchored" data-anchor-id="random-effects">Random effects</h5>
<p>Our final option takes a new approach. It considers the units we measured as simply a potential samples from a larger population. We use the information from the units we measured to consider the distribution of sample effects we might see (the variance we would expect among containers). For this to work, you probably want 5+ levels of the unit variable. This is because we are using the means to estimate variance (confusing?). For factors with &lt;5 levels, random effects likely offer no benefit <span class="citation" data-cites="gomes2022">(<a href="#ref-gomes2022" role="doc-biblioref">Gomes 2022</a>)</span>.</p>
<!--useful here: https://stats.oarc.ucla.edu/other/mult-pkg/introduction-to-linear-mixed-models/-->
<p>The impact of unit is then included in the model as a random effect<span class="citation" data-cites="introduca">(<a href="#ref-introduca" role="doc-biblioref"><span>“Introduction to Linear Mixed Models,”</span> n.d.</a>)</span>. When models contain fixed (what we’ve done before) and random effects, we call them mixed-effects models. To update our our model, we now have</p>
<p><span class="math display">\[
\begin{split}
Y=X\beta+ Z\mu+\epsilon , \textrm{ where }\\
\textrm{Y is our observations (an nx1 matrix)}\\
\textrm{X is a matrix showing our explanatory variables (an nxk matrix)}\\
\beta \textrm{ is our coefficient matrix(an kx1 matrix)}\\
Z  \textrm{ is a coefficient matrix that signifies each unit (an n x (n * number or random effects) matrix)}\\
\mu \textrm{ is our coefficient matrix for each unit (an (n * number of random effects) x 1 matrix) }\\
\epsilon \textrm{ is a matrix (an nx1) of residuals)}\\
\end{split}
\]</span></p>
<p>Two common packages for carrying out this analysis in R are the <strong>nlme</strong> and <strong>lme4</strong> packages. We will focus on the <strong>lme4</strong> package here. Random effects can be entered in the <em>lmer</em> (linear mixed-effects regression) function and specified as (1|Grouping Unit). One nice thing about <strong>lme4</strong> is it will handle crossed and random effects on it’s own <strong>as long as you don’t repeat unit names</strong>. For example, we could note</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lme4)</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>mixed_analysis <span class="ot">&lt;-</span> <span class="fu">lmer</span>(growth<span class="sc">~</span>predator_cue<span class="sc">+</span>(<span class="dv">1</span><span class="sc">|</span>tank), experiment_with_12_tanks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and we now assume</p>
<p><span class="math display">\[
\epsilon \approx i.i.d.\ N(\mu,\sigma)\\
\mu \approx N(0, \sigma_{\mu})
\]</span></p>
<p>The main assumption we add here is that the random effects are normally distributed. This means we can estimate all the random parameters at a single level (we will stick with one level for now) using just one degree of freedom.</p>
<p>Once we build these models,, we need to consider assumptions.This should be checked at each level of grouping.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(mixed_analysis)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-72-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Another package, <em>DHARMa</em> does this for models by considering quantile residuals, which should be flat/uniform when a model fits the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DHARMa)</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a><span class="fu">simulateResiduals</span>(mixed_analysis, <span class="at">n =</span> <span class="dv">1000</span>, <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-73-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of Class DHARMa with simulated residuals based on 1000 simulations with refit = FALSE . See ?DHARMa::simulateResiduals for help. 
 
Scaled residual values: 0.371 0.707 0.816 0.828 0.046 0.959 0.583 0.261 0.537 0.077 0.348 0.088 0.663 0.874 0.856 0.948 0.433 0.012 0.832 0.429 ...</code></pre>
</div>
</div>
<p>The only warning I would give with this package is that significant p-values may be mis-leading for large datasets,so just use a visual analysis (the package author has made this exact point in the documentation and in discussions <span class="citation" data-cites="r-gene">(<a href="#ref-r-gene" role="doc-biblioref"><span>“R - Generalised Linear Mixed Model Diagnostics Using DHARMa - Cross Validated,”</span> n.d.</a>)</span> ).</p>
<details>
<summary>
The old way
</summary>
<p>For simple models like this one, the <em>check_mixed_model</em> function (provided below) also offered an automated approach. This function was used before some of the newer packages used above were demonstrated.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>check_mixed_model <span class="ot">&lt;-</span> <span class="cf">function</span> (model, <span class="at">model_name =</span> <span class="cn">NULL</span>) {</span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>  <span class="co">#collection of things you might check for mixed model</span></span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))</span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">length</span>(<span class="fu">names</span>(<span class="fu">ranef</span>(model))<span class="sc">&lt;</span><span class="dv">2</span>)){</span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">qqnorm</span>(<span class="fu">ranef</span>(model, <span class="at">drop =</span> T)[[<span class="dv">1</span>]], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">las =</span> <span class="dv">1</span>, <span class="at">cex =</span> <span class="fl">1.4</span>, <span class="at">main=</span> <span class="fu">paste</span>(model_name, </span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a>                                                                                  <span class="st">"</span><span class="sc">\n</span><span class="st"> Random effects Q-Q plot"</span>))</span>
<span id="cb134-7"><a href="#cb134-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">qqline</span>(<span class="fu">ranef</span>(model, <span class="at">drop =</span> T)[[<span class="dv">1</span>]])</span>
<span id="cb134-8"><a href="#cb134-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb134-9"><a href="#cb134-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(<span class="fu">fitted</span>(model),<span class="fu">residuals</span>(model), <span class="at">main =</span> <span class="fu">paste</span>(model_name, </span>
<span id="cb134-10"><a href="#cb134-10" aria-hidden="true" tabindex="-1"></a>                                                    <span class="st">"</span><span class="sc">\n</span><span class="st"> residuals vs fitted"</span>))</span>
<span id="cb134-11"><a href="#cb134-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">qqnorm</span>(<span class="fu">residuals</span>(model), <span class="at">main =</span><span class="fu">paste</span>(model_name, </span>
<span id="cb134-12"><a href="#cb134-12" aria-hidden="true" tabindex="-1"></a>                                       <span class="st">"</span><span class="sc">\n</span><span class="st">residuals q-q plot"</span>))</span>
<span id="cb134-13"><a href="#cb134-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">qqline</span>(<span class="fu">residuals</span>(model))</span>
<span id="cb134-14"><a href="#cb134-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">hist</span>(<span class="fu">residuals</span>(model), <span class="at">main =</span> <span class="fu">paste</span>(model_name, </span>
<span id="cb134-15"><a href="#cb134-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="st">"</span><span class="sc">\n</span><span class="st">residual histogram"</span>))</span>
<span id="cb134-16"><a href="#cb134-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_mixed_model</span>(mixed_analysis)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-75-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This works well in that we can see the random effects and residuals are normally distributed.</p>
</details>
<p>We can now follow our normal protocol and consider the outcome using our <em>summary</em> command -</p>
<p>note the output denotes we have 225 observations and 9 grouping levels.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mixed_analysis)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Linear mixed model fit by REML ['lmerMod']
Formula: growth ~ predator_cue + (1 | tank)
   Data: experiment_with_12_tanks

REML criterion at convergence: -285.5

Scaled residuals: 
     Min       1Q   Median       3Q      Max 
-2.95584 -0.68320 -0.09694  0.73893  2.88985 

Random effects:
 Groups   Name        Variance Std.Dev.
 tank     (Intercept) 0.007214 0.08493 
 Residual             0.019931 0.14118 
Number of obs: 300, groups:  tank, 12

Fixed effects:
                 Estimate Std. Error t value
(Intercept)       5.15897    0.04475 115.278
predator_cuelow  -0.10021    0.06329  -1.583
predator_cuehigh -0.17006    0.06329  -2.687

Correlation of Fixed Effects:
            (Intr) prdtr_cl
predatr_clw -0.707         
predtr_chgh -0.707  0.500  </code></pre>
</div>
</div>
<p>We can also still use <em>Anova</em> to get p-values. However, these are now calculated by default using likelihood-associated <span class="math inline">\(\chi^2\)</span> tests.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(mixed_analysis, <span class="at">type =</span> <span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III Wald chisquare tests)

Response: growth
                 Chisq Df Pr(&gt;Chisq)    
(Intercept)  13289.026  1    &lt; 2e-16 ***
predator_cue     7.297  2    0.02603 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>You can also ask for F tests, but note the degrees of freedom associated with these tests is not clear. It’s somewhere between the “average” and “ignore” approach used above. You also will see these 2 approaches give different answers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(mixed_analysis, <span class="at">type =</span> <span class="st">"III"</span>, <span class="at">test=</span><span class="st">"F"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III Wald F tests with Kenward-Roger df)

Response: growth
                      F Df Df.res    Pr(&gt;F)    
(Intercept)  13289.0258  1      9 1.412e-15 ***
predator_cue     3.6485  2      9   0.06912 .  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">glht</span>(mixed_analysis, <span class="at">linfct =</span> <span class="fu">mcp</span>(<span class="st">"predator_cue"</span><span class="ot">=</span> <span class="st">"Tukey"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Simultaneous Tests for General Linear Hypotheses

Multiple Comparisons of Means: Tukey Contrasts


Fit: lmer(formula = growth ~ predator_cue + (1 | tank), data = experiment_with_12_tanks)

Linear Hypotheses:
                 Estimate Std. Error z value Pr(&gt;|z|)  
low - none == 0  -0.10021    0.06329  -1.583   0.2529  
high - none == 0 -0.17006    0.06329  -2.687   0.0197 *
high - low == 0  -0.06985    0.06329  -1.104   0.5118  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
(Adjusted p values reported -- single-step method)</code></pre>
</div>
</div>
<p>Note our results indicate we can only see differences between the high and control treatments. This is due to the noise we introduced in the data (which always happens in the real world) in addition to the random effect issues.</p>
<details>
<summary>
Is this really new?
</summary>
<p>It should be noted that random effects match up to paired t-tests and blocking approaches for simple experiments.Remember our feather experiment (example and data provided by <span class="citation" data-cites="mcdonald2014">McDonald (<a href="#ref-mcdonald2014" role="doc-biblioref">2014</a>)</span>)?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>two_way_anova_example <span class="ot">&lt;-</span> <span class="fu">lm</span>(Color_index <span class="sc">~</span> Feather_type <span class="sc">+</span> Bird, <span class="at">data=</span>feather)</span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(two_way_anova_example, <span class="at">type=</span> <span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: Color_index
              Sum Sq Df F value   Pr(&gt;F)    
(Intercept)  0.24133  1 26.5059 0.000119 ***
Feather_type 0.15043  1 16.5214 0.001017 ** 
Bird         0.21950 15  1.6072 0.184180    
Residuals    0.13657 15                     
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Note we can get the same results using a paired t-test, but (as of early 2024) we need to use a wide dataset.</p>
<!-- t.test no longer allowing paired in formula method as of early 2024  -->
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape2) </span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>feather_wide <span class="ot">&lt;-</span> <span class="fu">dcast</span>(feather, Bird<span class="sc">~</span>Feather_type, <span class="at">value.var=</span><span class="st">"Color_index"</span>)</span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="fu">paged_table</span>(feather_wide)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Bird"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Odd"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["Typical"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"A","2":"-0.324","3":"-0.255"},{"1":"B","2":"-0.185","3":"-0.213"},{"1":"C","2":"-0.299","3":"-0.190"},{"1":"D","2":"-0.144","3":"-0.185"},{"1":"E","2":"-0.027","3":"-0.045"},{"1":"F","2":"-0.039","3":"-0.025"},{"1":"G","2":"-0.264","3":"-0.015"},{"1":"H","2":"-0.077","3":"0.003"},{"1":"I","2":"-0.017","3":"0.015"},{"1":"J","2":"-0.169","3":"0.020"},{"1":"K","2":"-0.096","3":"0.023"},{"1":"L","2":"-0.330","3":"0.040"},{"1":"M","2":"-0.346","3":"0.040"},{"1":"N","2":"-0.191","3":"0.050"},{"1":"O","2":"-0.128","3":"0.055"},{"1":"P","2":"-0.182","3":"0.058"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Now we have wide data…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">t.test</span>(feather_wide<span class="sc">$</span>Odd, feather_wide<span class="sc">$</span>Typical, <span class="at">data=</span>feather, <span class="at">paired=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Paired t-test

data:  feather_wide$Odd and feather_wide$Typical
t = -4.0647, df = 15, p-value = 0.001017
alternative hypothesis: true mean difference is not equal to 0
95 percent confidence interval:
 -0.20903152 -0.06521848
sample estimates:
mean difference 
      -0.137125 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>two_way_lme_example <span class="ot">&lt;-</span> <span class="fu">lmer</span>(Color_index <span class="sc">~</span> Feather_type <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>Bird), <span class="at">data=</span>feather)</span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(two_way_lme_example, <span class="at">type=</span> <span class="st">"III"</span>, <span class="at">test=</span><span class="st">"F"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III Wald F tests with Kenward-Roger df)

Response: Color_index
                  F Df Df.res   Pr(&gt;F)    
(Intercept)  41.816  1 28.457 4.88e-07 ***
Feather_type 16.521  1 15.000 0.001017 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>All same if we use least-squares approaches (not the default <span class="math inline">\(\chi^2\)</span> tests that match likelihood-based methods).</p>
</details>
</section>
<section id="more-random-effects" class="level5">
<h5 class="anchored" data-anchor-id="more-random-effects">More random effects</h5>
<p>What we have demonstrated above is a <em>random intercepts</em> model. This means each group (tank in this case) differs slightly. This approach works well when we are considering impacts of categorical predictors (like an ANOVA). If we are considering slopes, we can also have <em>random slope</em> models, where every group responds slightly differently to the continuous predictor. You can also combine these into models with both random intercepts and random slopes.</p>
<!-- simulation @ https://verynormal.substack.com/p/office-hours-1-simulating-from-mixed -->
<p>As an example, consider work by <span class="citation" data-cites="oconnor2007">O’Connor et al. (<a href="#ref-oconnor2007" role="doc-biblioref">2007</a>)</span> on relationship between pelagic larval duration (pld) and temperature (exampled developed in <span class="citation" data-cites="midway">Midway (<a href="#ref-midway" role="doc-biblioref">n.d.</a>)</span>). The group collected data from 214 observations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a>pld <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/PLD.txt"</span>, <span class="at">stringsAsFactors =</span> T, <span class="at">header =</span> T)</span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pld, <span class="fu">aes</span>(<span class="at">y =</span> pld, <span class="at">x=</span> temp)) <span class="sc">+</span></span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">'Temperature (C)'</span>, </span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span><span class="st">'PLD (days)'</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-85-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>but they were all from 72 species. These 72 species could be averaged out …</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>averaged_species_pld <span class="ot">&lt;-</span> <span class="fu">summarySE</span>(pld, <span class="at">measurevar =</span> <span class="st">"pld"</span>,</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">groupvars =</span> <span class="fu">c</span>(<span class="st">"species"</span>))</span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a>averaged_species_temp <span class="ot">&lt;-</span> <span class="fu">summarySE</span>(pld, <span class="at">measurevar =</span> <span class="st">"temp"</span>,</span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">groupvars =</span> <span class="fu">c</span>(<span class="st">"species"</span>))</span>
<span id="cb152-5"><a href="#cb152-5" aria-hidden="true" tabindex="-1"></a>averaged_species <span class="ot">&lt;-</span> <span class="fu">merge</span>(averaged_species_pld[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>],averaged_species_temp[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">all =</span> T)</span>
<span id="cb152-6"><a href="#cb152-6" aria-hidden="true" tabindex="-1"></a><span class="fu">paged_table</span>(averaged_species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["species"],"name":[1],"type":["fct"],"align":["left"]},{"label":["N"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["pld"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["temp"],"name":[4],"type":["dbl"],"align":["right"]}],"data":[{"1":"Amphiprion.melanopus","2":"2","3":"10.650000","4":"26.50000"},{"1":"Armases.miersii","2":"2","3":"15.150000","4":"26.60000"},{"1":"Balanus.amphitrite.","2":"5","3":"11.000000","4":"22.00000"},{"1":"Balanus.eburneus","2":"3","3":"7.700000","4":"25.00000"},{"1":"Balanus.trigonus","2":"4","3":"6.500000","4":"22.75000"},{"1":"Callianassa.tyrrhena","2":"4","3":"8.125000","4":"19.75000"},{"1":"Cancer.gracilis","2":"3","3":"41.533333","4":"15.00000"},{"1":"Cancer.irroratus","2":"5","3":"25.800000","4":"17.60000"},{"1":"Cancer.magister.","2":"2","3":"53.700000","4":"12.50000"},{"1":"Cancer.oregonensis","2":"2","3":"53.250000","4":"12.50000"},{"1":"Cancer.productus","2":"3","3":"40.333333","4":"15.00000"},{"1":"Carcinus.maenas.","2":"4","3":"39.825000","4":"17.50000"},{"1":"Cardium.edule","2":"2","3":"25.000000","4":"17.50000"},{"1":"Chlamys.hastata","2":"2","3":"38.000000","4":"14.00000"},{"1":"Chthamalus.montagui","2":"3","3":"11.666667","4":"19.70000"},{"1":"Chthamalus.stellatus","2":"2","3":"15.500000","4":"20.75000"},{"1":"Circeus.spirillum","2":"3","3":"12.000000","4":"10.00000"},{"1":"Clupea.harengus.pallasi","2":"3","3":"5.000000","4":"8.00000"},{"1":"Crassostrea.virginica","2":"3","3":"15.333333","4":"25.50000"},{"1":"Crepidula.fornicata.","2":"2","3":"33.000000","4":"21.00000"},{"1":"Crepidula.plana","2":"4","3":"37.525000","4":"21.50000"},{"1":"Dendraster.excentricus","2":"3","3":"5.126667","4":"17.00000"},{"1":"Echinaster.type.I.","2":"3","3":"16.000000","4":"25.00000"},{"1":"Echinaster.type.II","2":"3","3":"17.333333","4":"25.00000"},{"1":"Elminius.modestus","2":"4","3":"18.750000","4":"15.00000"},{"1":"Eriocheir.sinensis.","2":"3","3":"66.500000","4":"15.00000"},{"1":"Gadus.morhua","2":"6","3":"34.000000","4":"9.00000"},{"1":"Haliotis.asinina","2":"3","3":"2.133333","4":"28.00000"},{"1":"Haliotis.fulgens","2":"2","3":"3.500000","4":"24.00000"},{"1":"Haliotis.sorenseni.","2":"5","3":"12.800000","4":"16.00000"},{"1":"Hemigrapsus.sanguineus","2":"3","3":"29.800000","4":"20.00000"},{"1":"Homarus.americanus","2":"5","3":"55.600000","4":"15.30000"},{"1":"Hyas.araneus","2":"3","3":"64.833333","4":"12.00000"},{"1":"Hydroides.elegans","2":"4","3":"5.675000","4":"22.50000"},{"1":"Laqueus.californianus","2":"3","3":"7.000000","4":"10.00000"},{"1":"Limulus.polyphemus","2":"3","3":"64.333333","4":"25.00000"},{"1":"Lithodes.santolla","2":"5","3":"56.940000","4":"9.00000"},{"1":"Lysmata.seticaudata","2":"2","3":"25.650000","4":"23.00000"},{"1":"Lytechinus.variegatus.","2":"2","3":"32.000000","4":"20.50000"},{"1":"Mactra.solidissima","2":"2","3":"27.000000","4":"18.00000"},{"1":"Menippe.mercenaria","2":"2","3":"34.500000","4":"27.50000"},{"1":"Mithrax.caribbaeus","2":"3","3":"11.666667","4":"25.00000"},{"1":"Mopalia.muscosa","2":"2","3":"23.250000","4":"13.50000"},{"1":"Mytilus.edulis","2":"2","3":"27.500000","4":"13.50000"},{"1":"Nassarius.obsoletus","2":"3","3":"15.666667","4":"21.90000"},{"1":"Necora.puber.","2":"3","3":"36.033333","4":"20.00000"},{"1":"Nereis.virens","2":"3","3":"9.666667","4":"10.66667"},{"1":"Ostrea.lurida","2":"2","3":"11.500000","4":"21.00000"},{"1":"Pagurus.bernhardus.","2":"3","3":"64.333333","4":"12.00000"},{"1":"Pagurus.criniticornis","2":"3","3":"20.266667","4":"25.00000"},{"1":"Palaemonetes.longirostris","2":"2","3":"19.750000","4":"22.50000"},{"1":"Palaemonetes.pugio","2":"3","3":"21.266667","4":"25.00000"},{"1":"Palaemonetes.varians","2":"3","3":"16.166667","4":"20.00000"},{"1":"Palaemonetes.vulgaris","2":"3","3":"20.833333","4":"25.43333"},{"1":"Pandalus.borealis","2":"2","3":"83.500000","4":"3.40000"},{"1":"Panopeus.herbstii","2":"2","3":"30.000000","4":"22.50000"},{"1":"Paralomis.granulosa","2":"5","3":"53.780000","4":"9.00000"},{"1":"Parthenope.serrata.","2":"2","3":"37.500000","4":"22.50000"},{"1":"Perna.viridis","2":"4","3":"23.775000","4":"27.75000"},{"1":"Pseudopleuronectes.americanus","2":"2","3":"64.500000","4":"6.50000"},{"1":"Rhithropanopeus.harrisii","2":"3","3":"21.333333","4":"25.00000"},{"1":"Scolecolepides.viridis","2":"2","3":"34.000000","4":"15.00000"},{"1":"Scylla.serrata","2":"4","3":"23.250000","4":"24.50000"},{"1":"Semibalanus.balanoides","2":"3","3":"14.000000","4":"12.00000"},{"1":"Sesarma.cinereum.","2":"3","3":"29.666667","4":"25.00000"},{"1":"Spirorbis.spirorbis","2":"3","3":"10.000000","4":"10.00000"},{"1":"Sprattus.sprattus","2":"2","3":"2.000000","4":"9.00000"},{"1":"Strombus.gigas","2":"3","3":"20.666667","4":"28.00000"},{"1":"Tivela.mactroides","2":"3","3":"16.866667","4":"25.00000"},{"1":"Tonicella.lineata","2":"2","3":"3.250000","4":"11.25000"},{"1":"Uca.pugilator","2":"3","3":"17.900000","4":"24.66667"},{"1":"Upeneus.tragula","2":"2","3":"10.600000","4":"27.50000"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>and plotted</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(averaged_species, <span class="fu">aes</span>(<span class="at">y =</span> pld, <span class="at">x=</span> temp)) <span class="sc">+</span></span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x=</span><span class="st">'Temperature (C)'</span>, </span>
<span id="cb153-4"><a href="#cb153-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">y=</span><span class="st">'PLD (days)'</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-87-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Models could be built according to both of these approaches. For now we will only consider a random intercept model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>averaged_model <span class="ot">&lt;-</span> <span class="fu">lm</span>(pld <span class="sc">~</span> temp, averaged_species)</span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a>mixed_model <span class="ot">&lt;-</span> <span class="fu">lmer</span>(pld <span class="sc">~</span> temp <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> species), pld)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and checked</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(averaged_model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-89-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-89-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-89-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-89-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(mixed_model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-89-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The average model looks ok, but the mixed-model is struggling with variances that are increasing over time. To follow the paper’s lead (despite what we wrote above), let’s log-transform the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>averaged_model_log <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="fu">log</span>(pld) <span class="sc">~</span> <span class="fu">log</span>(temp), averaged_species)</span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>mixed_model_log_intercept <span class="ot">&lt;-</span> <span class="fu">lmer</span>(<span class="fu">log</span>(pld) <span class="sc">~</span> <span class="fu">log</span>(temp) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> species), pld)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and check them again</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(averaged_model_log)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-91-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-91-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-91-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-91-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(mixed_model_log_intercept)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-91-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The mixed model looks better now. But how do these models impact our findings?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(averaged_model_log, <span class="at">type =</span> <span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: log(pld)
            Sum Sq Df F value    Pr(&gt;F)    
(Intercept) 25.024  1 36.8304 5.921e-08 ***
log(temp)    1.918  1  2.8232   0.09737 .  
Residuals   47.560 70                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(mixed_model_log_intercept, <span class="at">type =</span> <span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III Wald chisquare tests)

Response: log(pld)
              Chisq Df Pr(&gt;Chisq)    
(Intercept) 1247.28  1  &lt; 2.2e-16 ***
log(temp)    544.98  1  &lt; 2.2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>Here we see a clear benefit (if we trust p-values…) of including all our data. However, we might also think that each species has a unique relationship to temperature. This would be a random slopes models (notice we no longer have a constant intercept per(line) species, but instead have a 0.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>mixed_model_log_slope <span class="ot">&lt;-</span> <span class="fu">lmer</span>(<span class="fu">log</span>(pld) <span class="sc">~</span> <span class="fu">log</span>(temp) <span class="sc">+</span> (<span class="dv">0</span><span class="sc">+</span>temp <span class="sc">|</span> species), pld)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We could also have both</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>mixed_model_log_both <span class="ot">&lt;-</span> <span class="fu">lmer</span>(<span class="fu">log</span>(pld) <span class="sc">~</span> <span class="fu">log</span>(temp) <span class="sc">+</span> (temp <span class="sc">|</span> species), pld)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Which should we use? <span class="citation" data-cites="zuur2009">A. Zuur et al. (<a href="#ref-zuur2009" role="doc-biblioref">2009</a>)</span> recommends fitting a maximal model (all the fixed effects we might want to include, like starting at top-down approach) to the data and then comparing these using AIC.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paged_table</span>(<span class="fu">AICc</span>(mixed_model_log_both, mixed_model_log_intercept, mixed_model_log_slope))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["df"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["AICc"],"name":[2],"type":["dbl"],"align":["right"]}],"data":[{"1":"6","2":"198.0361","_rn_":"mixed_model_log_both"},{"1":"4","2":"229.0712","_rn_":"mixed_model_log_intercept"},{"1":"4","2":"418.7678","_rn_":"mixed_model_log_slope"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Here we find fitting both a random slope and intercept fit the data best. We can then analyze the model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(mixed_model_log_both, <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Deviance Table (Type III Wald chisquare tests)

Response: log(pld)
             Chisq Df Pr(&gt;Chisq)    
(Intercept) 650.19  1  &lt; 2.2e-16 ***
log(temp)   309.30  1  &lt; 2.2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>and note a significant impact of temperature on pelagic larval duration. This is another example of how blocking may have missed this, unless we considered interactions…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(<span class="fu">lm</span>(<span class="fu">log</span>(pld) <span class="sc">~</span> <span class="fu">log</span>(temp)<span class="sc">*</span>species, pld), <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: log(pld)
                  Sum Sq Df F value    Pr(&gt;F)    
(Intercept)       0.0775  1  3.8510   0.05369 .  
log(temp)         0.0488  1  2.4239   0.12401    
species           8.6095 71  6.0243 7.697e-13 ***
log(temp):species 4.4996 71  3.1485 1.511e-06 ***
Residuals         1.4090 70                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>which would have been harder to interpret given our lack of concern about species.</p>
</section>
</section>
<section id="consider-a-correlation-structure-the-bigger-picture" class="level4">
<h4 class="anchored" data-anchor-id="consider-a-correlation-structure-the-bigger-picture">Consider a correlation structure: The bigger picture</h4>
<p>Though not developed here (for now), random effects are an example of a correlation structure. We could also consider linkages among our data in different ways. This is common in spatial statistics and where time is involved. When data are connected in this way, various correlation structures can be included in the model.</p>
</section>
</section>
</section>
<section id="errors-are-not-equal-among-groups" class="level2">
<h2 class="anchored" data-anchor-id="errors-are-not-equal-among-groups">Errors are not equal among groups</h2>
<p>Issues with errors not being equal among groups is often tied to other issues, like non-linearity or the nature of the data, which we have addressed above. Correlation structures can also be used to address issues with errors are not equal among groups.</p>
<p>One example of this option is to use weighted-least squares regression (also known as generalized least squares) - this approach specifically helps when residuals are not evenly distributed among groups (or when a funnel/cone appears when you plot the model to check assumptions). For example, we could take the model on below-ground biomass that we developed in the <a href="../../content/chapters/More_ANOVAs.html">More Anovas chapter</a>. As a reminder, <span class="citation" data-cites="valdez2023">Valdez et al. (<a href="#ref-valdez2023" role="doc-biblioref">2023</a>)</span> wanted to consider the impact of top-down (snail grazing) and bottom- up (nutrient availability) on marsh plant (<em>Spartina alterniflora</em>) growth. To do this, they assigned plots to one of 3 grazer treatments and one of 2 nitrogen treatments. We focused on the impact of these treatments on standing dead mass and noted a funnel-shape when reviewing the residuals.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a>valdez_2023 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/Spartina_alterniflora_traits.csv"</span>, <span class="at">stringsAsFactors =</span> T)</span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a>sdm_model <span class="ot">&lt;-</span><span class="fu">lm</span>(Standing.Dead..dry..m2.<span class="sc">~</span>Snail.Level <span class="sc">*</span> Nitrogen.level, valdez_2023[valdez_2023<span class="sc">$</span>Snail.Level <span class="sc">!=</span> <span class="st">"uncaged"</span>,])</span>
<span id="cb171-3"><a href="#cb171-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sdm_model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-98-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-98-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-98-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-98-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>To address the, we can use weighted least squares. This approach assume you built the model and then noted an issue with heteroscedasticity. To use, we can calculate a weight for each residual that is based on its variance. Doing this requires an iterative process similar to those used for estimation for generalized linear models (we’ll get to these) - below makes a value that increases with low variance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a>wt_sdm <span class="ot">&lt;-</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fu">lm</span>(<span class="fu">abs</span>(sdm_model<span class="sc">$</span>residuals) <span class="sc">~</span> sdm_model<span class="sc">$</span>fitted.values)<span class="sc">$</span>fitted.values<span class="sc">^</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then add a new argument to the <em>lm</em> function to use these weights.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>sdm_model_wls <span class="ot">&lt;-</span><span class="fu">lm</span>(Standing.Dead..dry..m2.<span class="sc">~</span>Snail.Level <span class="sc">*</span> Nitrogen.level, valdez_2023[valdez_2023<span class="sc">$</span>Snail.Level <span class="sc">!=</span> <span class="st">"uncaged"</span>,], <span class="at">weights =</span> wt_sdm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then continue on our normal route:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sdm_model_wls) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-101-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-101-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-101-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-101-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Anova</span>(sdm_model_wls, <span class="at">type=</span><span class="st">"III"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Anova Table (Type III tests)

Response: Standing.Dead..dry..m2.
                            Sum Sq Df F value    Pr(&gt;F)    
(Intercept)                203.228  1  76.442 1.497e-06 ***
Snail.Level                157.447  2  29.611 2.288e-05 ***
Nitrogen.level             130.416  1  49.054 1.427e-05 ***
Snail.Level:Nitrogen.level 143.013  2  26.896 3.681e-05 ***
Residuals                   31.903 12                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>If you compare the two models you notice slight differences - these are minimal here due to lack of differences in variance.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sdm_model) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Standing.Dead..dry..m2. ~ Snail.Level * Nitrogen.level, 
    data = valdez_2023[valdez_2023$Snail.Level != "uncaged", 
        ])

Residuals:
    Min      1Q  Median      3Q     Max 
-3.5800 -0.8350 -0.0317  0.6600  3.7400 

Coefficients:
                                                Estimate Std. Error t value
(Intercept)                                       18.710      1.032  18.124
Snail.Levelremoval                               -14.150      1.460  -9.692
Snail.Levelsnail addition                         -7.567      1.460  -5.183
Nitrogen.levelwithout                            -15.240      1.460 -10.439
Snail.Levelremoval:Nitrogen.levelwithout          16.153      2.065   7.824
Snail.Levelsnail addition:Nitrogen.levelwithout   10.357      2.065   5.016
                                                Pr(&gt;|t|)    
(Intercept)                                     4.39e-10 ***
Snail.Levelremoval                              5.02e-07 ***
Snail.Levelsnail addition                       0.000228 ***
Nitrogen.levelwithout                           2.25e-07 ***
Snail.Levelremoval:Nitrogen.levelwithout        4.72e-06 ***
Snail.Levelsnail addition:Nitrogen.levelwithout 0.000301 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.788 on 12 degrees of freedom
Multiple R-squared:  0.9284,    Adjusted R-squared:  0.8986 
F-statistic: 31.14 on 5 and 12 DF,  p-value: 1.792e-06</code></pre>
</div>
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sdm_model_wls)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Standing.Dead..dry..m2. ~ Snail.Level * Nitrogen.level, 
    data = valdez_2023[valdez_2023$Snail.Level != "uncaged", 
        ], weights = wt_sdm)

Weighted Residuals:
     Min       1Q   Median       3Q      Max 
-2.31784 -0.89013 -0.03702  0.92266  2.46121 

Coefficients:
                                                Estimate Std. Error t value
(Intercept)                                       18.710      2.140   8.743
Snail.Levelremoval                               -14.150      2.202  -6.426
Snail.Levelsnail addition                         -7.567      2.490  -3.039
Nitrogen.levelwithout                            -15.240      2.176  -7.004
Snail.Levelremoval:Nitrogen.levelwithout          16.153      2.322   6.956
Snail.Levelsnail addition:Nitrogen.levelwithout   10.357      2.620   3.953
                                                Pr(&gt;|t|)    
(Intercept)                                     1.50e-06 ***
Snail.Levelremoval                              3.27e-05 ***
Snail.Levelsnail addition                        0.01030 *  
Nitrogen.levelwithout                           1.43e-05 ***
Snail.Levelremoval:Nitrogen.levelwithout        1.53e-05 ***
Snail.Levelsnail addition:Nitrogen.levelwithout  0.00192 ** 
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.631 on 12 degrees of freedom
Multiple R-squared:  0.8747,    Adjusted R-squared:  0.8225 
F-statistic: 16.75 on 5 and 12 DF,  p-value: 4.789e-05</code></pre>
</div>
</div>
<p>Why not just always do this? Because weighted least squares implicitly assumes we <em>know</em> the weights. We are actually estimating them, so small datasets may lead to bad estimates and outcomes.</p>
<p>For another example (from <span class="citation" data-cites="zuur2009">A. Zuur et al. (<a href="#ref-zuur2009" role="doc-biblioref">2009</a>)</span> from <span class="citation" data-cites="smith2005">Smith et al. (<a href="#ref-smith2005" role="doc-biblioref">2005</a>)</span> ), we could consider the seasonal relationship between the testis weight and dorsal mantle length of the squid <em>Loligo forbesii</em>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://animalia.bio/loligo-forbesii"><img src="../..\images/loligo-forbesii.webp" class="img-fluid figure-img" alt="Restored lithograph of Loligo forbesii. Public domain."></a></p>
<figcaption>Restored lithograph of Loligo forbesii. Public domain.</figcaption>
</figure>
</div>
<p>A basic model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a>squid <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/Squid.txt"</span>, <span class="at">stringsAsFactors =</span> T, <span class="at">header =</span> T)</span>
<span id="cb181-2"><a href="#cb181-2" aria-hidden="true" tabindex="-1"></a>squid<span class="sc">$</span>MONTH <span class="ot">&lt;-</span> <span class="fu">factor</span>(squid<span class="sc">$</span>MONTH)</span>
<span id="cb181-3"><a href="#cb181-3" aria-hidden="true" tabindex="-1"></a>squid_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(Testisweight<span class="sc">~</span>DML<span class="sc">*</span>MONTH, squid)</span>
<span id="cb181-4"><a href="#cb181-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(squid_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-103-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-103-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-103-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-103-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Shows some issues. We see some over-dispersion,but more importantly there is a funnel-shape to the data. This means variance is increasing with mean. We can also see this in</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_model</span>(squid_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-104-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="fu">simulateResiduals</span>(squid_lm, <span class="at">n =</span> <span class="dv">1000</span>, <span class="at">plot =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-104-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Object of Class DHARMa with simulated residuals based on 1000 simulations with refit = FALSE . See ?DHARMa::simulateResiduals for help. 
 
Scaled residual values: 0.253 0.791 0.485 0.596 0.759 0.772 0.204 0.738 0.704 0.748 0.817 0.635 0.926 0.836 0.331 0.843 0.654 0.579 0.483 0.667 ...</code></pre>
</div>
</div>
<p>Unlike when we measured the same organism or unit multiple times, this time the reason for variation is clear - some months have more spread than other</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(squid, <span class="fu">aes</span>(<span class="at">x=</span> MONTH,  <span class="at">y=</span>Testisweight))<span class="sc">+</span></span>
<span id="cb185-2"><a href="#cb185-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-105-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>and this is seen in residual structure</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb186"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb186-1"><a href="#cb186-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(squid_lm, <span class="fu">aes</span>(<span class="at">x =</span> MONTH, <span class="at">y =</span> .resid)) <span class="sc">+</span></span>
<span id="cb186-2"><a href="#cb186-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb186-3"><a href="#cb186-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Linear_model_extensions_files/figure-html/unnamed-chunk-106-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Instead of assuming the residuals are identically distributed (we see they are not), we can instead specify the variance structure. We have numerous options here. A fixed variance sets a formula for the variance</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(nlme)</span>
<span id="cb187-2"><a href="#cb187-2" aria-hidden="true" tabindex="-1"></a>squid_fixed <span class="ot">&lt;-</span> <span class="fu">gls</span>(Testisweight<span class="sc">~</span>DML<span class="sc">*</span>MONTH, squid, <span class="at">weights =</span> <span class="fu">varFixed</span>(<span class="sc">~</span>DML))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>which allows for it to vary (here, with DML). We can also let variance be based on another factor (here, month)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a>squid_month <span class="ot">&lt;-</span> <span class="fu">gls</span>(Testisweight<span class="sc">~</span>DML<span class="sc">*</span>MONTH, squid, <span class="at">weights =</span> <span class="fu">varIdent</span>(<span class="sc">~</span><span class="dv">1</span><span class="sc">|</span>MONTH))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Other options exist (<em>varPower</em>, <em>varExp</em>, <em>varConstPower</em>, and <em>varComb</em>. Models can be compared using AIC</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb189"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" aria-hidden="true" tabindex="-1"></a><span class="fu">AICc</span>(squid_fixed,squid_month, squid_lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            df     AICc
squid_fixed 25 3622.650
squid_month 25 3753.836
squid_lm    25 3643.629</code></pre>
</div>
</div>
<table class="table">
<caption>Correlation structure options</caption>
<colgroup>
<col style="width: 25%">
<col style="width: 75%">
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Use</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>VarFixed</td>
<td>Fixed variance - constant or multiplier/equation option</td>
</tr>
<tr class="even">
<td>VarIdent</td>
<td>Variance differs across levels of some factor</td>
</tr>
<tr class="odd">
<td>VarPower</td>
<td>Variance relates to some power (estimated) of another variable (similar to VarFixed)</td>
</tr>
<tr class="even">
<td>VarExp</td>
<td>Variance proportional to e<sup>variable</sup></td>
</tr>
<tr class="odd">
<td>VarConstPower</td>
<td>Variance relates to a constant plus some variable <sup>constant</sup></td>
</tr>
<tr class="even">
<td>VarComb</td>
<td>Variance proportional to e<sup>variable&nbsp;*&nbsp;constant</sup> (combines VarIdent and varExp)</td>
</tr>
</tbody>
</table>
<p>Models using this approach should show no structure in plots of normalized residuals (raw residual divided by square root of the variance).</p>
</section>
<section id="residuals-are-not-normally-distributed" class="level2">
<h2 class="anchored" data-anchor-id="residuals-are-not-normally-distributed">Residuals are not normally distributed</h2>
<p>A very common concern regarding linear models is normality. I list it last here because this assumption is one of the least important (and the assumption is based on residuals, not data!). However, non-normal residuals are often (not always) connected to other issues, namely linearity, as noted above.</p>
</section>
<section id="next-steps-combining-these-approaches" class="level2">
<h2 class="anchored" data-anchor-id="next-steps-combining-these-approaches">Next steps: Combining these approaches</h2>
<p>Many of these approaches can be combined. For example, random effects be included in generalized linear models and generalized additive models , or a model could include random effects for measurement unit and a correlation structure given other relationships. Random effects could also be included in a model focused on transformed variables. It can all get really complicated - the key is to focus on making the model fit the data and ensuring assumptions are met before considering results. As noted at the beginning, this module is introducing you to these techniques so you are prepared to explore them further. Suggested reading includes <span class="citation" data-cites="zuur2009">A. Zuur et al. (<a href="#ref-zuur2009" role="doc-biblioref">2009</a>)</span>, <span class="citation" data-cites="harrison2018">Harrison et al. (<a href="#ref-harrison2018" role="doc-biblioref">2018</a>)</span>, <span class="citation" data-cites="corlatti2021">Corlatti (<a href="#ref-corlatti2021" role="doc-biblioref">2021</a>)</span>, and <span class="citation" data-cites="crawley2014">Crawley (<a href="#ref-crawley2014" role="doc-biblioref">2014</a>)</span>.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-agresti2007" class="csl-entry" role="listitem">
Agresti, Alan. 2007. <span>“Introduction to Categorical Data Analysis.”</span>
</div>
<div id="ref-applied2022" class="csl-entry" role="listitem">
<span>“Applied Regression Analysis and Generalized Linear Models.”</span> 2022. <a href="https://us.sagepub.com/en-us/nam/applied-regression-analysis-and-generalized-linear-models/book237254">https://us.sagepub.com/en-us/nam/applied-regression-analysis-and-generalized-linear-models/book237254</a>.
</div>
<div id="ref-blais2023" class="csl-entry" role="listitem">
Blais, Brian R., Samantha L. Johnson, and John L. Koprowski. 2023. <span>“Effects of Disturbances and Environmental Changes on an Aridland Riparian Generalist.”</span> <em>PeerJ</em> 11 (June): e15563. <a href="https://doi.org/10.7717/peerj.15563">https://doi.org/10.7717/peerj.15563</a>.
</div>
<div id="ref-blanca2017" class="csl-entry" role="listitem">
Blanca, María J., Rafael Alarcón, Jaume Arnau, Roser Bono, and Rebecca Bendayan. 2017. <span>“Non-normal data: Is ANOVA still a valid option?”</span> <em>Psicothema</em> 29 (4): 552–57. <a href="https://doi.org/10.7334/psicothema2016.383">https://doi.org/10.7334/psicothema2016.383</a>.
</div>
<div id="ref-blanca2018" class="csl-entry" role="listitem">
———. 2018. <span>“Effect of Variance Ratio on ANOVA Robustness: Might 1.5 Be the Limit?”</span> <em>Behavior Research Methods</em> 50 (3): 937–62. <a href="https://doi.org/10.3758/s13428-017-0918-2">https://doi.org/10.3758/s13428-017-0918-2</a>.
</div>
<div id="ref-bolker2008" class="csl-entry" role="listitem">
Bolker, Benjamin M. 2008. <em>Ecological Models and Data in r</em>. 508th edition. Princeton: Princeton University Press.
</div>
<div id="ref-bueno2020" class="csl-entry" role="listitem">
Bueno, Marilia, Glauco B. O. Machado, and Fosca P. P. Leite. 2020. <span>“Colonization of Novel Algal Habitats by Juveniles of a Marine Tube-Dwelling Amphipod.”</span> <em>PeerJ</em> 8 (October): e10188. <a href="https://doi.org/10.7717/peerj.10188">https://doi.org/10.7717/peerj.10188</a>.
</div>
<div id="ref-corlatti2021" class="csl-entry" role="listitem">
Corlatti, Luca. 2021. <span>“Regression Models, Fantastic Beasts, and Where to Find Them: A Simple Tutorial for Ecologists Using R.”</span> <em>Bioinformatics and Biology Insights</em>, October. <a href="https://doi.org/10.1177/11779322211051522">https://doi.org/10.1177/11779322211051522</a>.
</div>
<div id="ref-crawley2014" class="csl-entry" role="listitem">
Crawley, Michael J. 2014. <em>Statistics: An Introduction Using r</em>. 2nd edition. Chichester, West Sussex, UK: Wiley.
</div>
<div id="ref-dornberger2016" class="csl-entry" role="listitem">
Dornberger, Lindsey, Cameron Ainsworth, Stephen Gosnell, and Felicia Coleman. 2016. <span>“Developing a Polycyclic Aromatic Hydrocarbon Exposure Dose-Response Model for Fish Health and Growth.”</span> <em>Marine Pollution Bulletin</em> 109 (1): 259–66. <a href="https://doi.org/10.1016/j.marpolbul.2016.05.072">https://doi.org/10.1016/j.marpolbul.2016.05.072</a>.
</div>
<div id="ref-feng2020" class="csl-entry" role="listitem">
Feng, Cindy, Longhai Li, and Alireza Sadeghpour. 2020. <span>“A Comparison of Residual Diagnosis Tools for Diagnosing Regression Models for Count Data.”</span> <em>BMC Medical Research Methodology</em> 20 (1): 175. <a href="https://doi.org/10.1186/s12874-020-01055-2">https://doi.org/10.1186/s12874-020-01055-2</a>.
</div>
<div id="ref-fox2015" class="csl-entry" role="listitem">
Fox, John. 2015. <em>Applied Regression Analysis and Generalized Linear Models</em>. SAGE Publications.
</div>
<div id="ref-fox2016" class="csl-entry" role="listitem">
———. 2016. <em>Applied Regression Analysis and Generalized Linear Models</em>. Third Edition. Los Angeles: SAGE.
</div>
<div id="ref-geissinger2022" class="csl-entry" role="listitem">
Geissinger, Emilie A., Celyn L. L. Khoo, Isabella C. Richmond, Sally J. M. Faulkner, and David C. Schneider. 2022. <span>“A Case for Beta Regression in the Natural Sciences.”</span> <em>Ecosphere</em> 13 (2): e3940. <a href="https://doi.org/10.1002/ecs2.3940">https://doi.org/10.1002/ecs2.3940</a>.
</div>
<div id="ref-gelman2006" class="csl-entry" role="listitem">
Gelman, Andrew, and Jennifer Hill. 2006. <em>Data Analysis Using Regression and Multilevel/Hierarchical Models</em>. 1st edition. Cambridge ; New York: Cambridge University Press.
</div>
<div id="ref-gomes2022" class="csl-entry" role="listitem">
Gomes, Dylan G. E. 2022. <span>“Should I Use Fixed Effects or Random Effects When I Have Fewer Than Five Levels of a Grouping Factor in a Mixed-Effects Model?”</span> <em>PeerJ</em> 10 (January): e12794. <a href="https://doi.org/10.7717/peerj.12794">https://doi.org/10.7717/peerj.12794</a>.
</div>
<div id="ref-gosnell2023" class="csl-entry" role="listitem">
Gosnell, J. Stephen, Xochitl S. Clare, Dillon Whited, Jennifer Chiu, Stephen Huie, Danielle Zacherl, and Crow White. 2023. <span>“Length-Weight Relationship of the Kelp Forest Gastropod and Emerging Fisheries Species, Kellet’s Whelk, Kelletia Kelletii.”</span> <em>Journal of Shellfish Research</em> 42 (3). <a href="https://doi.org/10.2983/035.042.0313">https://doi.org/10.2983/035.042.0313</a>.
</div>
<div id="ref-harrison2018" class="csl-entry" role="listitem">
Harrison, Xavier A., Lynda Donaldson, Maria Eugenia Correa-Cano, Julian Evans, David N. Fisher, Cecily E. D. Goodwin, Beth S. Robinson, David J. Hodgson, and Richard Inger. 2018. <span>“A Brief Introduction to Mixed Effects Modelling and Multi-Model Inference in Ecology.”</span> <em>PeerJ</em> 6 (May): e4794. <a href="https://doi.org/10.7717/peerj.4794">https://doi.org/10.7717/peerj.4794</a>.
</div>
<div id="ref-heffner1996" class="csl-entry" role="listitem">
Heffner, Robert A., Mark J. Butler, and Colleen Keelan Reilly. 1996. <span>“Pseudoreplication Revisited.”</span> <em>Ecology</em> 77 (8): 2558–62. <a href="https://doi.org/10.2307/2265754">https://doi.org/10.2307/2265754</a>.
</div>
<div id="ref-hunt" class="csl-entry" role="listitem">
Hunt, Michael. n.d. <span>“QQ-Plots.”</span> <em>RPubs</em>. <a href="https://rpubs.com/mbh038/725314">https://rpubs.com/mbh038/725314</a>.
</div>
<div id="ref-hurlbert1984" class="csl-entry" role="listitem">
Hurlbert, Stuart H. 1984. <span>“Pseudoreplication and the Design of Ecological Field Experiments.”</span> <em>Ecological Monographs</em> 54 (2): 187–211. <a href="https://doi.org/10.2307/1942661">https://doi.org/10.2307/1942661</a>.
</div>
<div id="ref-introduca" class="csl-entry" role="listitem">
<span>“Introduction to Linear Mixed Models.”</span> n.d. <a href="https://stats.oarc.ucla.edu/other/mult-pkg/introduction-to-linear-mixed-models/">https://stats.oarc.ucla.edu/other/mult-pkg/introduction-to-linear-mixed-models/</a>.
</div>
<div id="ref-knief2021" class="csl-entry" role="listitem">
Knief, Ulrich, and Wolfgang Forstmeier. 2021. <span>“Violating the Normality Assumption May Be the Lesser of Two Evils.”</span> <em>Behavior Research Methods</em> 53 (6): 2576–90. <a href="https://doi.org/10.3758/s13428-021-01587-5">https://doi.org/10.3758/s13428-021-01587-5</a>.
</div>
<div id="ref-lazic2010" class="csl-entry" role="listitem">
Lazic, Stanley E. 2010. <span>“The Problem of Pseudoreplication in Neuroscientific Studies: Is It Affecting Your Analysis?”</span> <em>BMC Neuroscience</em> 11 (1): 5. <a href="https://doi.org/10.1186/1471-2202-11-5">https://doi.org/10.1186/1471-2202-11-5</a>.
</div>
<div id="ref-li2024" class="csl-entry" role="listitem">
Li, Dongbo, Benjamin Brough, Jasper W. Rees, Christophe F. D. Coste, Chenggui Yuan, Mike S. Fowler, and Steven M. Sait. 2024. <span>“Humidity Modifies Species-Specific and Age-Dependent Heat Stress Effects in an Insect Host-Parasitoid Interaction.”</span> <em>Ecology and Evolution</em> 14 (7): e70047. <a href="https://doi.org/10.1002/ece3.70047">https://doi.org/10.1002/ece3.70047</a>.
</div>
<div id="ref-lindelov" class="csl-entry" role="listitem">
Lindelov, J. K. n.d. <em>Common Statistical Tests Are Linear Models (or: How to Teach Stats)</em>. <a href="https://lindeloev.github.io/tests-as-linear/">https://lindeloev.github.io/tests-as-linear/</a>.
</div>
<div id="ref-mcdonald2014" class="csl-entry" role="listitem">
McDonald, J. H. 2014. <em>Handbook of Biological Statistics</em>. 3rd ed. Baltimore, MD: Sparky House Publishing.
</div>
<div id="ref-midway" class="csl-entry" role="listitem">
Midway, Steve. n.d. <em>Data Analysis in r</em>. <a href="https://bookdown.org/steve_midway/DAR/">https://bookdown.org/steve_midway/DAR/</a>.
</div>
<div id="ref-molenberghs2012" class="csl-entry" role="listitem">
Molenberghs, Geert, Geert Verbeke, Samuel Iddi, and Clarice G. B. Demétrio. 2012. <span>“A Combined Beta and Normal Random-Effects Model for Repeated, Overdispersed Binary and Binomial Data.”</span> <em>Journal of Multivariate Analysis</em> 111 (October): 94–109. <a href="https://doi.org/10.1016/j.jmva.2012.05.005">https://doi.org/10.1016/j.jmva.2012.05.005</a>.
</div>
<div id="ref-montgomery2012" class="csl-entry" role="listitem">
Montgomery, Douglas C., Elizabeth A. Peck, and G. Geoffrey Vining. 2012. <em>Introduction to Linear Regression Analysis</em>. 5th edition. Hoboken, NJ: Wiley.
</div>
<div id="ref-oconnor2007" class="csl-entry" role="listitem">
O’Connor, M. I., J. F. Bruno, S. D. Gaines, B. S. Halpern, S. E. Lester, B. P. Kinlan, and J. M. Weiss. 2007. <span>“Temperature Control of Larval Dispersal and the Implications for Marine Ecology, Evolution, and Conservation.”</span> <em>Proceedings of the National Academy of Sciences of the United States of America</em> 104 (4): 1266–71. <a href="internal-pdf://O_Connoretal2007[1]-3537220354/O_Connoretal2007[1].pdf">internal-pdf://O_Connoretal2007[1]-3537220354/O_Connoretal2007[1].pdf</a>.
</div>
<div id="ref-payne2017" class="csl-entry" role="listitem">
Payne, Elizabeth H., James W. Hardin, Leonard E. Egede, Viswanathan Ramakrishnan, Anbesaw Selassie, and Mulugeta Gebregziabher. 2017. <span>“Approaches for dealing with various sources of overdispersion in modeling count data: Scale adjustment versus modeling.”</span> <em>Statistical Methods in Medical Research</em> 26 (4): 1802–23. <a href="https://doi.org/10.1177/0962280215588569">https://doi.org/10.1177/0962280215588569</a>.
</div>
<div id="ref-r-gene" class="csl-entry" role="listitem">
<span>“R - Generalised Linear Mixed Model Diagnostics Using DHARMa - Cross Validated.”</span> n.d. <a href="https://stats.stackexchange.com/questions/478369/generalised-linear-mixed-model-diagnostics-using-dharma">https://stats.stackexchange.com/questions/478369/generalised-linear-mixed-model-diagnostics-using-dharma</a>.
</div>
<div id="ref-reitan2016" class="csl-entry" role="listitem">
Reitan, Trond, and Anders Nielsen. 2016. <span>“Do Not Divide Count Data with Count Data; A Story from Pollination Ecology with Implications Beyond.”</span> <em>PLOS ONE</em> 11 (2): e0149129. <a href="https://doi.org/10.1371/journal.pone.0149129">https://doi.org/10.1371/journal.pone.0149129</a>.
</div>
<div id="ref-series" class="csl-entry" role="listitem">
Series, Developed, and maintained by the contributors of the QCBS R. Workshop. n.d. <em>Workshop 6: Generalized Linear Models</em>. <a href="https://r.qcbs.ca/workshop06/book-en/index.html">https://r.qcbs.ca/workshop06/book-en/index.html</a>.
</div>
<div id="ref-shatz2023" class="csl-entry" role="listitem">
Shatz, Itamar. 2023. <span>“Assumption-Checking Rather Than (Just) Testing: The Importance of Visualization and Effect Size in Statistical Diagnostics.”</span> <em>Behavior Research Methods</em>, March. <a href="https://doi.org/10.3758/s13428-023-02072-x">https://doi.org/10.3758/s13428-023-02072-x</a>.
</div>
<div id="ref-smith2005" class="csl-entry" role="listitem">
Smith, Jennifer, Graham Pierce, Alain Zuur, and Peter Boyle. 2005. <span>“Seasonal Patterns of Investment in Reproductive and Somatic Tissues in the Squid Loligo Forbesi.”</span> <em>Aquatic Living Resources - AQUAT LIVING RESOUR</em> 18 (October): 341–51. <a href="https://doi.org/10.1051/alr:2005038">https://doi.org/10.1051/alr:2005038</a>.
</div>
<div id="ref-valdez2023" class="csl-entry" role="listitem">
Valdez, Stephanie R., Pedro Daleo, David S. DeLaMater Iii, and Brian R. Silliman. 2023. <span>“Variable Responses to Top-down and Bottom-up Control on Multiple Traits in the Foundational Plant, Spartina Alterniflora.”</span> <em>PLOS ONE</em> 18 (5): e0286327. <a href="https://doi.org/10.1371/journal.pone.0286327">https://doi.org/10.1371/journal.pone.0286327</a>.
</div>
<div id="ref-warton2016" class="csl-entry" role="listitem">
Warton, David I., Mitchell Lyons, Jakub Stoklosa, and Anthony R. Ives. 2016. <span>“Three Points to Consider When Choosing a LM or GLM Test for Count Data.”</span> <em>Methods in Ecology and Evolution</em> 7 (8): 882–90. <a href="https://doi.org/10.1111/2041-210X.12552">https://doi.org/10.1111/2041-210X.12552</a>.
</div>
<div id="ref-warton2010" class="csl-entry" role="listitem">
Warton, David, and Francis Hui. 2010. <span>“The Arcsine Is Asinine: The Analysis of Proportions in Ecology.”</span> <em>Ecology</em> 92 (1): 3–10. <a href="https://doi.org/10.1890/10-0340.1">https://doi.org/10.1890/10-0340.1</a>.
</div>
<div id="ref-workshop" class="csl-entry" role="listitem">
<em>Workshop 6: Generalized Linear Models</em>. n.d. <a href="https://r.qcbs.ca/workshop06/book-en/index.html">https://r.qcbs.ca/workshop06/book-en/index.html</a>.
</div>
<div id="ref-zuur2010" class="csl-entry" role="listitem">
Zuur, Alain F., Elena N. Ieno, and Chris S. Elphick. 2010. <span>“A Protocol for Data Exploration to Avoid Common Statistical Problems.”</span> <em>Methods in Ecology and Evolution</em> 1 (1): 3–14. https://doi.org/<a href="https://doi.org/10.1111/j.2041-210X.2009.00001.x">https://doi.org/10.1111/j.2041-210X.2009.00001.x</a>.
</div>
<div id="ref-zuur2007" class="csl-entry" role="listitem">
Zuur, Alain, Elena N. Ieno, and Graham M. Smith. 2007. <em>Analyzing Ecological Data</em>. 2007th edition. New York ; London: Springer.
</div>
<div id="ref-zuur2009" class="csl-entry" role="listitem">
Zuur, Alain, Elena N. Ieno, Neil Walker, Anatoly A. Saveliev, and Graham M. Smith. 2009. <em>Mixed Effects Models and Extensions in Ecology with r</em>. New York, NY: Springer.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/jsgosnell\.github\.io\/cuny_biostats_book\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../content/chapters/Combining_numerical_and_categorical_predictors.html" class="pagination-link  aria-label=" combining="" numerical="" and="" categorical="" predictors"="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Combining numerical and categorical predictors</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../content/chapters/multivariate_methods.html" class="pagination-link" aria-label="Multivariate methods">
        <span class="nav-page-text">Multivariate methods</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en">© CC BY-NC-SA 4.0</a></p>
</div>   
    <div class="nav-footer-center">
<p><a href="https://jsgosnell.github.io/cuny_biostats_book/">Course notes</a> &amp; <a href="https://jsgosnell.github.io/cuny_biostats_website/">slides</a></p>
<div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/jsgosnell/cuny_biostats_book/edit/main/content/chapters/Linear_model_extensions.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/jsgosnell/cuny_biostats_book/blob/main/content/chapters/Linear_model_extensions.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li><li><a href="https://github.com/jsgosnell/cuny_biostats_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This page is built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>